% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{book}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Darwin Core Guide},
  pdfauthor={Standardizing Marine Biological Data Working Group},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\usepackage{booktabs}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi

\title{Darwin Core Guide}
\author{\href{https://github.com/ioos/bio_data_guide/graphs/contributors}{Standardizing Marine Biological Data Working Group}}
\date{2022-02-10}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}
\addcontentsline{toc}{chapter}{Preface}

Biological data structures, definitions, measurements, and linkages are neccessarily as diverse as the systems they represent. This presents a real challenge when integrating data across biological research domains such as ecology, oceanography, fisheries, and climate sciences.

Lots of standards exist for use with biological data but navigating them can be difficult for data managers who are new to them. The Earth Science Information Partners (ESIP) Biological Data Standards Cluster developed \href{https://doi.org/10.6084/m9.figshare.16806712.v1}{this primer} for managers of biological data to provide a quick, easy resource for navigating a selection of the standards that exist. The goal of the primer is to spread awareness about existing standards and is intended to be shared online and at conferences to increase the adoption of standards for biological data and make them FAIR.

Benson, Abigail; LaScala-Gruenewald, Diana; McGuinn, Robert; Satterthwaite, Erin; Beaulieu, Stace; Biddle, Mathew; et al.~(2021): Biological Observation Data Standardization - A Primer for Data Managers. ESIP. Online resource. \url{https://doi.org/10.6084/m9.figshare.16806712.v1}

\hypertarget{intro}{%
\chapter{Introduction}\label{intro}}

The world of standardizing marine biological data can seem complex for the naive oceanographer, biologist, scientist, or programmer.
Transforming and integrating data is about combining the right standards for your desired interoperability with other data types.
For example, interoperating fish biology measurements with climate level variables.
There are a few concepts necessary to make this possible such as standard data structures, controlled vocabularies and knowledge representations, along with metadata standards to facilitate data discovery. This will permit the inclusion of more data and broader access to better ecosystem based models. Many scientific domains data handling practices are currently being reshaped in light of recent advances in computing power, technology, and data science.

\hypertarget{data-structures}{%
\section{Data Structures}\label{data-structures}}

The OBIS-ENV Darwin Core Archive Data Structure.

\href{https://obis.org/manual/}{OBIS manual}

\hypertarget{ontologies}{%
\section{Ontologies}\label{ontologies}}

An ontology is a classification system for establishing a hierarchically related set of concepts. Concepts are often terms from controlled vocabularies.

From Marine Metadata: \# TODO: add link

``Ontologies can include all of the following, but are not required to include them, depending on which perspective from above you adhere to:

Classes (general things, types of things)
Instances (individual things)
Relationships among things
Properties of things
Functions, processes, constraints, and rules relating to things''

TODO: Research Unified Modelling Language?

\href{http://www.obofoundry.org/ontology/envo.html}{Environment Ontology (EnvO)} EnvO is a community ontology for the concise, controlled description of environments.

\hypertarget{controlled-vocabularies}{%
\section{Controlled Vocabularies}\label{controlled-vocabularies}}

There are a number of controlled vocabularies that are used to describe parameters commonly used in specific research domains. This allows for greater interoperability of data sets within the domain, and ideally between domains. Here, we strive to document a number of relevant examples.

\begin{itemize}
\item
  \href{http://cfconventions.org/standard-names.html}{Climate and Format (CF) Standard Names} The purpose of the standard\_name attribute is to provide a succinct and distinguishing description of a variable, in a way that encourages interoperability. These terms are typically for physical observations, however, there have been advancements in aligning biological taxa into the CF standard names (see \href{https://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html\#taxon-names-and-identifiers}{here}).
\item
  \href{http://vocab.nerc.ac.uk/}{NERC Vocabulary Server} The NVS gives access to standardised and hierarchically-organized vocabularies.

  \begin{itemize}
  \item
    \href{http://vocab.nerc.ac.uk/collection/L05/current/}{Device categories} using the SeaDataNet device categories
  \item
    \href{http://vocab.nerc.ac.uk/collection/L22/current/}{Device make/model using the SeaVoX Device Catalogue}
  \item
    \href{http://vocab.nerc.ac.uk/collection/L06/current/}{Platform categories using SeaVoX Platform Categories}
  \item
    \href{http://vocab.nerc.ac.uk/collection/C17/current/}{Platform instances using the ICES Platform Codes}
  \item
    \href{http://vocab.nerc.ac.uk/collection/P06/current/}{Unit of measure}
  \end{itemize}
\item
  \href{https://wiki.earthdata.nasa.gov/display/CMR/GCMD+Keyword+Access}{GCMD Keywords (NASA)}
\item
  \href{http://vocab.nerc.ac.uk/collection/C19/current/}{Geographic Domain/Features of Interest}
\item
  \href{http://schema.geolink.org/1.0/base/main.html}{GeoLink base ontology} was part of the \href{http://www.geolink.org/}{EarthCube GeoLink Project}
\end{itemize}

Note: To describe a measurement or fact of a biological specimen that conforms to Darwin Core standards, it's necessary to use the `Biological entity described elsewhere' method rather than taxon specific.

\hypertarget{taxonomy}{%
\subsection{Taxonomy}\label{taxonomy}}

\begin{itemize}
\tightlist
\item
  \href{https://www.marinespecies.org/}{The World Registry of Marine Species (WoRMS)} The aim of a World Register of Marine Species (WoRMS) is to provide an authoritative and comprehensive list of names of marine organisms, including information on synonymy. While the highest priority goes to valid names, other names in use are included so that this register can serve as a guide to interpret taxonomic literature.
\end{itemize}

\hypertarget{resources}{%
\subsection{Resources}\label{resources}}

\hypertarget{oceanography}{%
\subsection{Oceanography}\label{oceanography}}

\href{http://www.bco-dmo.org/}{Biological and Chemcial Oceanography Data Management Office}

\href{https://mmisw.org/ont/\#/}{Marine metadata interoperability vocab resources}

\hypertarget{biology}{%
\subsection{Biology}\label{biology}}

\href{http://bioportal.bioontology.org/ontologies/ECSO}{BioPortal Ecosystem Ontology}

\hypertarget{nerc-search-interfaces}{%
\subsection{NERC Search Interfaces}\label{nerc-search-interfaces}}

\begin{itemize}
\item
  \href{http://seadatanet.maris2.nl/v_bodc_vocab_v2/welcome.asp}{SeaDatanet Common Vocab Search Interface:}
\item
  \href{https://www.seadatanet.org/Standards/Common-Vocabularies/}{SeaDataNet Common Vocabularies:}
\item
  \href{http://seadatanet.maris2.nl/v_bodc_vocab_v2/vocab_relations.asp?lib=P08}{SeaDataNet Vocab Library}
\item
  \href{https://mof.obis.org/}{Measurement Types in OBIS}
\end{itemize}

\hypertarget{geosciences}{%
\subsection{Geosciences}\label{geosciences}}

\href{https://www.unidata.ucar.edu/software/udunits/}{UDUNITS}are more common unit measurements in geosciences

\hypertarget{ecoenvo}{%
\subsection{Eco/EnvO}\label{ecoenvo}}

\href{http://www.obofoundry.org/ontology/envo.html}{Environment Ontology} including genomics.

\hypertarget{wild-cards}{%
\subsection{Wild Cards}\label{wild-cards}}

Question: Not sure use case for this.

\href{https://www.bodc.ac.uk/resources/vocabularies/vocabulary_builder/biomodel/}{P01 Biological Entity Parameter Code Builder}

\hypertarget{technologies}{%
\section{Technologies}\label{technologies}}

\hypertarget{erddap}{%
\subsection{ERDDAP}\label{erddap}}

\href{https://coastwatch.pfeg.noaa.gov/erddap/index.html}{ERDDAP} can be thought of as a data server. It provides `easier access to scientific data' by providing a consistent interface that aggregates many disparate data sources. It does this by providing translation services between many common file types for gridded arrarys (`net CDF' files) and tabular data (spreadsheets). Data access is also made easier because it unifies different types of data servers and access protocols. \href{https://github.com/HakaiInstitute/erddap-basic}{Here} is a basic erddap installation that walks you through how to load a data set.

\hypertarget{notes-on-integrating-obis-darwin-core-as-it-relates-to-ooss}{%
\section{Notes on Integrating OBIS, Darwin Core as it relates to OOS's}\label{notes-on-integrating-obis-darwin-core-as-it-relates-to-ooss}}

\hypertarget{metadata}{%
\section{Metadata}\label{metadata}}

OBIS uses the \href{http://rs.gbif.org/schema/eml-gbif-profile/1.1/eml-gbif-profile.xsd}{GBIF EML profile} (version 1.1). In case data providers use ISO19115/ISO19139, there is a mapping available here: \url{http://rs.gbif.org/schema/eml-gbif-profile/1.1/eml2iso19139.xsl} This will be important for integrating OBIS datasets to other CIOOS and IOOS metadata profiles.

\hypertarget{data-qc}{%
\section{Data QC}\label{data-qc}}

There are a number of tools available to check the quality of data or check your data format against the expected standard.

\href{https://obis.org/manual/processing/}{OBIS Datatools} shows some great R packages for this.

\hypertarget{compliance-checking}{%
\subsection{Compliance Checking}\label{compliance-checking}}

LifeWatch Belgium provides a number of tools to check your data against.
Specifically you can test OBIS data format and see a map of your sample locations to check if they are on land.
See \url{http://www.lifewatch.be/data-services/}

There's also the \href{https://www.gbif.org/tools/data-validator}{GBIF data validator} which allows anyone with a GBIF-relevant dataset to receive a report on the syntactical correctness and the validity of the content contained within the dataset.

\hypertarget{semantic-web-and-darwin-core}{%
\subsection{Semantic Web and Darwin Core}\label{semantic-web-and-darwin-core}}

\href{http://www.semantic-web-journal.net/system/files/swj1093.pdf}{Lessons learned from adapting the Darwin Core vocabulary standard for use in RDF}

\hypertarget{resource-description-framework}{%
\subsection{Resource Description Framework}\label{resource-description-framework}}

\href{https://dwc.tdwg.org/rdf/}{Darwin Core Resource Description Framework Guide}

\hypertarget{applications}{%
\chapter{Applications}\label{applications}}

Some applications are demonstrated in this chapter.

\hypertarget{salmon-ocean-ecology-data}{%
\section{Salmon Ocean Ecology Data}\label{salmon-ocean-ecology-data}}

\hypertarget{intro-1}{%
\subsection{Intro}\label{intro-1}}

One of the goals of the Hakai Institute and the Canadian Integrated Ocean Observing System (CIOOS) is to facilitate Open Science and FAIR (findable, accessible, interoperable, reusable) ecological and oceanographic data. In a concerted effort to adopt or establish how best to do that, several Hakai and CIOOS staff attended an International Ocean Observing System (IOOS) Code Sprint in Ann Arbour, Michigan between October 7--11, 2019, to discuss how to implement FAIR data principles for biological data collected in the marine environment.

The \href{https://dwc.tdwg.org}{Darwin Core} is a highly structured data format that standardizes data table relations, vocabularies, and defines field names. The Darwin Core defines three table types: \texttt{event}, \texttt{occurrence}, and \texttt{measurementOrFact}. This intuitively captures the way most ecologists conduct their research. Typically, a survey (event) is conducted and measurements, counts, or observations (collectively measurementOrFacts) are made regarding a specific habitat or species (occurrence).

In the following script I demonstrate how I go about converting a subset of the data collected from the Hakai Institute Juvenile Salmon Program and discuss challenges, solutions, pros and cons, and when and what's worthwhile to convert to Darwin Core.

The conversion of a dataset to Darwin Core is much easier if your data are already tidy (normalized) in which you represent your data in separate tables that reflect the hierarchical and related nature of your observations. If your data are not already in a consistent and structured format, the conversion would likely be very arduos and not intuitive.

\hypertarget{event}{%
\subsection{event}\label{event}}

The first step is to consider what you will define as an event in your data set. I defined the capture of fish using a purse seine net as the \texttt{event}. Therefore, each row in the \texttt{event} table is one deployment of a seine net and is assigned a unique \texttt{eventID}.

My process for conversion was to make a new table called \texttt{event} and map the standard Darwin Core column names to pre-existing columns that serve the same purpose in my original \texttt{seine\_data} table and populate the other required fields.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event }\OtherTok{\textless{}{-}} \FunctionTok{tibble}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{seine\_id,}
                \AttributeTok{eventDate =} \FunctionTok{date}\NormalTok{(survey\_seines}\SpecialCharTok{$}\NormalTok{survey\_date),}
                \AttributeTok{decimalLatitude =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{lat,}
                \AttributeTok{decimalLongitude =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{long,}
                \AttributeTok{geodeticDatum =} \StringTok{"EPSG:4326 WGS84"}\NormalTok{,}
                \AttributeTok{minimumDepthInMeters =} \DecValTok{0}\NormalTok{,}
                \AttributeTok{maximumDepthInMeters =} \DecValTok{9}\NormalTok{, }\CommentTok{\# seine depth is 9 m}
                \AttributeTok{samplingProtocol =} \StringTok{"http://dx.doi.org/10.21966/1.566666"} \CommentTok{\# This is the DOI for the Hakai Salmon Data Package that contains the smnpling protocol, as well as the complete data package}
\NormalTok{               ) }

\FunctionTok{write\_csv}\NormalTok{(event, here}\SpecialCharTok{::}\FunctionTok{here}\NormalTok{(}\StringTok{"datasets"}\NormalTok{, }\StringTok{"hakai\_salmon\_data"}\NormalTok{, }\StringTok{"event.csv"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{occurrence}{%
\subsection{occurrence}\label{occurrence}}

Next you'll want to determine what constitutes an occurrence for your data set. Because each event caputers fish, I consider each fish to be an occurrence. Therefore, the unit of observation (each row) in the occurrence table is a fish. To link each occurence to an event you need to include the \texttt{eventID} column for every occurrence so that you know what seine (event) each fish (occurrence) came from. You must also provide a globally unique identifier for each occurrence. I already have a locally unique identifier for each fish in the original \texttt{fish\_data} table called \texttt{ufn}. To make it globally unique I pre-pend the organization and research program metadata to the \texttt{ufn} column.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\#TODO: Include bycatch data as well}

\DocumentationTok{\#\# make table long first}
\NormalTok{seines\_total\_long }\OtherTok{\textless{}{-}}\NormalTok{ survey\_seines }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(seine\_id, so\_total, pi\_total, cu\_total, co\_total, he\_total, ck\_total) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_longer}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{seine\_id, }\AttributeTok{names\_to =} \StringTok{"scientificName"}\NormalTok{, }\AttributeTok{values\_to =} \StringTok{"n"}\NormalTok{)}

\NormalTok{seines\_total\_long}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{recode}\NormalTok{(seines\_total\_long}\SpecialCharTok{$}\NormalTok{scientificName, }\AttributeTok{so\_total =} \StringTok{"Oncorhynchus nerka"}\NormalTok{, }\AttributeTok{pi\_total =} \StringTok{"Oncorhynchus gorbushca"}\NormalTok{, }\AttributeTok{cu\_total =} \StringTok{"Oncorhynchus keta"}\NormalTok{, }\AttributeTok{co\_total =} \StringTok{"Oncorhynchus kisutch"}\NormalTok{, }\AttributeTok{ck\_total =} \StringTok{"Oncorhynchus tshawytscha"}\NormalTok{, }\AttributeTok{he\_total =} \StringTok{"Clupea pallasii"}\NormalTok{) }

\NormalTok{seines\_taken\_long }\OtherTok{\textless{}{-}}\NormalTok{ survey\_seines }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(seine\_id, so\_taken, pi\_taken, cu\_taken, co\_taken, he\_taken, ck\_taken) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_longer}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{seine\_id, }\AttributeTok{names\_to =} \StringTok{"scientificName"}\NormalTok{, }\AttributeTok{values\_to =} \StringTok{"n\_taken"}\NormalTok{) }

\NormalTok{seines\_taken\_long}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{recode}\NormalTok{(seines\_taken\_long}\SpecialCharTok{$}\NormalTok{scientificName, }\AttributeTok{so\_taken =} \StringTok{"Oncorhynchus nerka"}\NormalTok{, }\AttributeTok{pi\_taken =} \StringTok{"Oncorhynchus gorbushca"}\NormalTok{, }\AttributeTok{cu\_taken =} \StringTok{"Oncorhynchus keta"}\NormalTok{, }\AttributeTok{co\_taken =} \StringTok{"Oncorhynchus kisutch"}\NormalTok{, }\AttributeTok{ck\_taken =} \StringTok{"Oncorhynchus tshawytscha"}\NormalTok{, }\AttributeTok{he\_taken =} \StringTok{"Clupea pallasii"}\NormalTok{) }

\DocumentationTok{\#\# remove records that have already been assigned an ID  }
\NormalTok{seines\_long }\OtherTok{\textless{}{-}}  \FunctionTok{full\_join}\NormalTok{(seines\_total\_long, seines\_taken\_long, }\AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"seine\_id"}\NormalTok{, }\StringTok{"scientificName"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{drop\_na}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{n\_not\_taken =}\NormalTok{ n }\SpecialCharTok{{-}}\NormalTok{ n\_taken) }\SpecialCharTok{\%\textgreater{}\%} \CommentTok{\#so\_total includes the number taken so I subtract n\_taken to get n\_not\_taken}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{n\_taken, }\SpecialCharTok{{-}}\NormalTok{n) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{filter}\NormalTok{(n\_not\_taken }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{)}

\NormalTok{all\_fish\_caught }\OtherTok{\textless{}{-}}
\NormalTok{  seines\_long[}\FunctionTok{rep}\NormalTok{(}\FunctionTok{seq.int}\NormalTok{(}\DecValTok{1}\NormalTok{, }\FunctionTok{nrow}\NormalTok{(seines\_long)), seines\_long}\SpecialCharTok{$}\NormalTok{n\_not\_taken), }\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{] }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{n\_not\_taken) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{prefix =} \StringTok{"hakai{-}jsp{-}"}\NormalTok{,}
         \AttributeTok{suffix =} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{nrow}\NormalTok{(.),}
         \AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(prefix, suffix)}
\NormalTok{  ) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{prefix, }\SpecialCharTok{{-}}\NormalTok{suffix)}

\CommentTok{\#}

\CommentTok{\# Change species names to full Scientific names }
\NormalTok{latin }\OtherTok{\textless{}{-}} \FunctionTok{fct\_recode}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{species, }\StringTok{"Oncorhynchus nerka"} \OtherTok{=} \StringTok{"SO"}\NormalTok{, }\StringTok{"Oncorhynchus gorbuscha"} \OtherTok{=} \StringTok{"PI"}\NormalTok{, }\StringTok{"Oncorhynchus keta"} \OtherTok{=} \StringTok{"CU"}\NormalTok{, }\StringTok{"Oncorhynchus kisutch"} \OtherTok{=} \StringTok{"CO"}\NormalTok{, }\StringTok{"Clupea pallasii"} \OtherTok{=} \StringTok{"HE"}\NormalTok{, }\StringTok{"Oncorhynchus tshawytscha"} \OtherTok{=} \StringTok{"CK"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{as.character}\NormalTok{()}

\NormalTok{fish\_retained\_data }\OtherTok{\textless{}{-}}\NormalTok{ fish\_data }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{scientificName =}\NormalTok{ latin) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{species) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{prefix =} \StringTok{"hakai{-}jsp{-}"}\NormalTok{,}
         \AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(prefix, ufn)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{semsp\_id, }\SpecialCharTok{{-}}\NormalTok{prefix, }\SpecialCharTok{{-}}\NormalTok{ufn, }\SpecialCharTok{{-}}\NormalTok{fork\_length\_field, }\SpecialCharTok{{-}}\NormalTok{fork\_length, }\SpecialCharTok{{-}}\NormalTok{weight, }\SpecialCharTok{{-}}\NormalTok{weight\_field)}

\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(all\_fish\_caught, fish\_retained\_data) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{basisOfRecord =} \StringTok{"HumanObservation"}\NormalTok{,}
        \AttributeTok{occurenceStatus =} \StringTok{"present"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ seine\_id)}
\end{Highlighting}
\end{Shaded}

For each occuerence of the six different fish species that I caught I need to match the species name that I provide with the official \texttt{scientificName} that is part of the World Register of Marine Species database \url{http://www.marinespecies.org/}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# I went directly to the WoRMS webite (http://www.marinespecies.org/) to download the full taxonomic levels for the salmon species I have and put the WoRMS output (species\_matched.xls) table in this project directory which is read in below and joined with the occurrence table}

\NormalTok{species\_matched }\OtherTok{\textless{}{-}}\NormalTok{ readxl}\SpecialCharTok{::}\FunctionTok{read\_excel}\NormalTok{(here}\SpecialCharTok{::}\FunctionTok{here}\NormalTok{(}\StringTok{"datasets"}\NormalTok{, }\StringTok{"hakai\_salmon\_data"}\NormalTok{, }\StringTok{"raw\_data"}\NormalTok{, }\StringTok{"species\_matched.xls"}\NormalTok{))}

\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(occurrence, species\_matched, }\AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"scientificName"} \OtherTok{=} \StringTok{"ScientificName"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(occurrenceID, basisOfRecord, scientificName, eventID, }\AttributeTok{occurrenceStatus =}\NormalTok{ occurenceStatus, Kingdom, Phylum, Class, Order, Family, Genus, Species)}

\FunctionTok{write\_csv}\NormalTok{(occurrence, here}\SpecialCharTok{::}\FunctionTok{here}\NormalTok{(}\StringTok{"datasets"}\NormalTok{, }\StringTok{"hakai\_salmon\_data"}\NormalTok{, }\StringTok{"occurrence.csv"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{measurementorfact}{%
\subsection{measurementOrFact}\label{measurementorfact}}

To convert all your measurements or facts from your normal format to Darwin Core you essentially need to put all your measurements into one column called measurementType and a corresponding column called MeasurementValue. This standardizes the column names are in the \texttt{measurementOrFact} table. There are a number of predefined \texttt{measurementType}s listed on the \href{https://www.bodc.ac.uk/resources/vocabularies/}{NERC} database that should be used where possible. I found it difficult to navigate this page to find the correct \texttt{measurementType}.

Here I convert length, and weight measurements that relate to an event and an occurrence and call those \texttt{measurementTypes} as \texttt{length} and \texttt{weight}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fish\_data}\SpecialCharTok{$}\NormalTok{weight }\OtherTok{\textless{}{-}} \FunctionTok{coalesce}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{weight, fish\_data}\SpecialCharTok{$}\NormalTok{weight\_field)}
\NormalTok{fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length }\OtherTok{\textless{}{-}} \FunctionTok{coalesce}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length, fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length\_field)}

\NormalTok{fish\_length }\OtherTok{\textless{}{-}}\NormalTok{ fish\_data }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"hakai{-}jsp{-}"}\NormalTok{, ufn)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(occurrenceID, }\AttributeTok{eventID =}\NormalTok{ seine\_id, fork\_length, weight) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{measurementType =} \StringTok{"fork length"}\NormalTok{, }\AttributeTok{measurementValue =}\NormalTok{ fork\_length) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(eventID, occurrenceID, measurementType, measurementValue) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{measurementUnit =} \StringTok{"millimeters"}\NormalTok{,}
         \AttributeTok{measurementUnitID =} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/UXMM/"}\NormalTok{)}

\NormalTok{fish\_weight }\OtherTok{\textless{}{-}}\NormalTok{ fish\_data }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"hakai{-}jsp{-}"}\NormalTok{, ufn)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(occurrenceID, }\AttributeTok{eventID =}\NormalTok{ seine\_id, fork\_length, weight) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{measurementType =} \StringTok{"mass"}\NormalTok{, }\AttributeTok{measurementValue =}\NormalTok{ weight) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(eventID, occurrenceID, measurementType, measurementValue) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{measurementUnit =} \StringTok{"grams"}\NormalTok{,}
         \AttributeTok{measurementUnitID =} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/UGRM/"}\NormalTok{)}

\NormalTok{measurementOrFact }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(fish\_length, fish\_weight) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{drop\_na}\NormalTok{(measurementValue)}

\FunctionTok{rm}\NormalTok{(fish\_length, fish\_weight)}

\FunctionTok{write\_csv}\NormalTok{(measurementOrFact, here}\SpecialCharTok{::}\FunctionTok{here}\NormalTok{(}\StringTok{"datasets"}\NormalTok{, }\StringTok{"hakai\_salmon\_data"}\NormalTok{, }\StringTok{"measurementOrFact.csv"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{hakai-seagrass}{%
\section{Hakai Seagrass}\label{hakai-seagrass}}

\hypertarget{setup}{%
\subsection{Setup}\label{setup}}

This section clears the workspace, checks the working directory, and
installs packages (if required) and loads packages, and loads necessary
datasets

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(}\StringTok{"knitr"}\NormalTok{)}
\CommentTok{\# Knitr global chunk options}
\NormalTok{opts\_chunk}\SpecialCharTok{$}\FunctionTok{set}\NormalTok{(}\AttributeTok{message =} \ConstantTok{FALSE}\NormalTok{,}
               \AttributeTok{warning =} \ConstantTok{FALSE}\NormalTok{,}
               \AttributeTok{error   =} \ConstantTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{load-data}{%
\subsubsection{Load Data}\label{load-data}}

First load the seagrass density survey data, set variable classes, and have a quick look

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load density data}
\NormalTok{seagrassDensity }\OtherTok{\textless{}{-}} 
  \FunctionTok{read.csv}\NormalTok{(}\StringTok{"raw\_data/seagrass\_density\_survey.csv"}\NormalTok{,}
           \AttributeTok{colClass =} \StringTok{"character"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{date             =} \FunctionTok{ymd}\NormalTok{(date),}
         \AttributeTok{depth            =} \FunctionTok{as.numeric}\NormalTok{(depth),}
         \AttributeTok{transect\_dist    =} \FunctionTok{factor}\NormalTok{(transect\_dist),}
         \AttributeTok{collected\_start  =} \FunctionTok{ymd\_hms}\NormalTok{(collected\_start),}
         \AttributeTok{collected\_end    =} \FunctionTok{ymd\_hms}\NormalTok{(collected\_end),}
         \AttributeTok{density          =} \FunctionTok{as.numeric}\NormalTok{(density),}
         \AttributeTok{density\_msq      =} \FunctionTok{as.numeric}\NormalTok{(density\_msq),}
         \AttributeTok{canopy\_height\_cm =} \FunctionTok{as.numeric}\NormalTok{(canopy\_height\_cm),}
         \AttributeTok{flowering\_shoots =} \FunctionTok{as.numeric}\NormalTok{(flowering\_shoots)) }\SpecialCharTok{\%T\textgreater{}\%}
  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 3,031
## Columns: 22
## $ X                <chr> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "1~
## $ organization     <chr> "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI",~
## $ work_area        <chr> "CALVERT", "CALVERT", "CALVERT", "CALVERT", "CALVERT"~
## $ project          <chr> "MARINEGEO", "MARINEGEO", "MARINEGEO", "MARINEGEO", "~
## $ survey           <chr> "PRUTH_BAY", "PRUTH_BAY", "PRUTH_BAY", "PRUTH_BAY", "~
## $ site_id          <chr> "PRUTH_BAY_INTERIOR4", "PRUTH_BAY_INTERIOR4", "PRUTH_~
## $ date             <date> 2016-05-13, 2016-05-13, 2016-05-13, 2016-05-13, 2016~
## $ sampling_bout    <chr> "4", "4", "4", "4", "4", "4", "4", "6", "6", "6", "6"~
## $ dive_supervisor  <chr> "Zach", "Zach", "Zach", "Zach", "Zach", "Zach", "Zach~
## $ collector        <chr> "Derek", "Derek", "Derek", "Derek", "Derek", "Derek",~
## $ hakai_id         <chr> "2016-05-13_PRUTH_BAY_INTERIOR4_0", "2016-05-13_PRUTH~
## $ sample_type      <chr> "seagrass_density", "seagrass_density", "seagrass_den~
## $ depth            <dbl> 6.0, 6.0, 6.0, 6.0, 5.0, 6.0, 6.0, 9.1, 9.0, 8.9, 9.0~
## $ transect_dist    <fct> 0, 5, 10, 15, 20, 25, 30, 10, 15, 20, 25, 30, 0, 5, 1~
## $ collected_start  <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ collected_end    <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ density          <dbl> 13, 10, 18, 22, 16, 31, 9, 5, 6, 6, 6, 3, 13, 30, 23,~
## $ density_msq      <dbl> 208, 160, 288, 352, 256, 496, 144, 80, 96, 96, 96, 48~
## $ canopy_height_cm <dbl> 60, 63, 80, 54, 55, 50, 63, 85, 80, 90, 95, 75, 60, 6~
## $ flowering_shoots <dbl> NA, NA, NA, NA, NA, NA, NA, 0, 0, 0, 0, 0, NA, NA, NA~
## $ comments         <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ quality_log      <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
\end{verbatim}

Next, load the habitat survey data, and same as above, set variable classes as necessary,
and have a quick look.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# load habitat data, set variable classes, have a quick look}
\NormalTok{seagrassHabitat }\OtherTok{\textless{}{-}}
  \FunctionTok{read.csv}\NormalTok{(}\StringTok{"raw\_data/seagrass\_habitat\_survey.csv"}\NormalTok{,}
           \AttributeTok{colClasses =} \StringTok{"character"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}  
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{date            =} \FunctionTok{ymd}\NormalTok{(date),}
         \AttributeTok{depth           =} \FunctionTok{as.numeric}\NormalTok{(depth),}
         \AttributeTok{hakai\_id        =} \FunctionTok{str\_pad}\NormalTok{(hakai\_id, }\DecValTok{5}\NormalTok{, }\AttributeTok{pad =} \StringTok{"0"}\NormalTok{),}
         \AttributeTok{transect\_dist   =} \FunctionTok{factor}\NormalTok{(transect\_dist),}
         \AttributeTok{collected\_start =} \FunctionTok{ymd\_hms}\NormalTok{(collected\_start),}
         \AttributeTok{collected\_end   =} \FunctionTok{ymd\_hms}\NormalTok{(collected\_end)) }\SpecialCharTok{\%T\textgreater{}\%}
  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 2,052
## Columns: 28
## $ X                <chr> "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "1~
## $ organization     <chr> "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI",~
## $ work_area        <chr> "CALVERT", "CALVERT", "CALVERT", "CALVERT", "CALVERT"~
## $ project          <chr> "MARINEGEO", "MARINEGEO", "MARINEGEO", "MARINEGEO", "~
## $ survey           <chr> "CHOKED_PASS", "CHOKED_PASS", "CHOKED_PASS", "CHOKED_~
## $ site_id          <chr> "CHOKED_PASS_INTERIOR6", "CHOKED_PASS_INTERIOR6", "CH~
## $ date             <date> 2017-11-22, 2017-11-22, 2017-11-22, 2017-11-22, 2017~
## $ sampling_bout    <chr> "6", "6", "6", "6", "6", "6", "1", "1", "1", "1", "1"~
## $ dive_supervisor  <chr> "gillian", "gillian", "gillian", "gillian", "gillian"~
## $ collector        <chr> "zach", "zach", "zach", "zach", "zach", "zach", "kyle~
## $ hakai_id         <chr> "10883", "2017-11-22_CHOKED_PASS_INTERIOR6_5 - 10", "~
## $ sample_type      <chr> "seagrass_habitat", "seagrass_habitat", "seagrass_hab~
## $ depth            <dbl> 9.2, 9.4, 9.3, 9.0, 9.2, 9.2, 3.4, 3.4, 3.4, 3.4, 3.4~
## $ transect_dist    <fct> 0 - 5, 10-May, 15-Oct, 15 - 20, 20 - 25, 25 - 30, 0 -~
## $ collected_start  <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ collected_end    <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ bag_uid          <chr> "10883", NA, NA, "11094", NA, "11182", "7119", NA, "7~
## $ bag_number       <chr> "3557", NA, NA, "3520", NA, "903", "800", NA, "318", ~
## $ density_range    <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ substrate        <chr> "sand,shell hash", "sand,shell hash", "sand,shell has~
## $ patchiness       <chr> "< 1", "< 1", "02-Jan", "< 1", "< 1", "< 1", "< 1", "~
## $ adj_habitat_1    <chr> "seagrass", "seagrass", "seagrass", "seagrass", "seag~
## $ adj_habitat_2    <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ sample_collected <chr> "TRUE", "FALSE", "FALSE", "TRUE", "FALSE", "TRUE", "T~
## $ vegetation_1     <chr> NA, NA, NA, NA, NA, NA, "des", NA, "des", NA, NA, NA,~
## $ vegetation_2     <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ comments         <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ quality_log      <chr> "1: Flowering shoots 0 for entire transects", NA, NA,~
\end{verbatim}

Finally, load coordinate data for surveys, and subset necessary variables

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{coordinates }\OtherTok{\textless{}{-}} 
  \FunctionTok{read.csv}\NormalTok{(}\StringTok{"raw\_data/seagrassCoordinates.csv"}\NormalTok{,}
           \AttributeTok{colClass =} \FunctionTok{c}\NormalTok{(}\StringTok{"Point.Name"} \OtherTok{=} \StringTok{"character"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(Point.Name, Decimal.Lat, Decimal.Long) }\SpecialCharTok{\%T\textgreater{}\%}
  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 70
## Columns: 3
## $ Point.Name   <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ Decimal.Lat  <dbl> 52.06200, 52.05200, 51.92270, 51.92500, 51.80900, 51.8090~
## $ Decimal.Long <dbl> -128.4120, -128.4030, -128.4648, -128.4540, -128.2360, -1~
\end{verbatim}

\hypertarget{merge-datasets}{%
\subsubsection{Merge Datasets}\label{merge-datasets}}

Now all the datasets have been loaded, and briefly formatted, we'll join
together the habitat and density surveys, and the coordinates for these.

The seagrass density surveys collect data at discrete points (ie. 5 metres)
along the transects, while the habitat surveys collect data over sections
(ie. 0 - 5 metres) along the transects. In order to fit these two surveys
together, we'll narrow the habitat surveys from a range to a point so the
locations will match. Based on how the habitat data is collected, the point
the habitat survey is applied to will be the distance at the end of the
swath (ie. 10-15m will become 15m). To account for no preceeding distance,
the 0m distance will use the 0-5m section of the survey.

First, well make the necessary transformations to the habitat dataset.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Reformat seagrassHabitat to merge with seagrassDensity}
\DocumentationTok{\#\# replicate 0 {-} 5m transect dist to match with 0m in density survey;}
\DocumentationTok{\#\# rest of habitat bins can map one to one with density (ie. 5 {-} 10m {-}\textgreater{} 10m)}
\NormalTok{seagrass0tmp }\OtherTok{\textless{}{-}} 
\NormalTok{  seagrassHabitat }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(transect\_dist }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"0 {-} 5"}\NormalTok{, }\StringTok{"0 {-} 2.5"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{transect\_dist =} \FunctionTok{factor}\NormalTok{(}\DecValTok{0}\NormalTok{))}

\DocumentationTok{\#\# collapse various levels to match with seagrassDensity transect\_dist}
\NormalTok{seagrassHabitat}\SpecialCharTok{$}\NormalTok{transect\_dist }\OtherTok{\textless{}{-}} 
  \FunctionTok{fct\_collapse}\NormalTok{(seagrassHabitat}\SpecialCharTok{$}\NormalTok{transect\_dist,}
               \StringTok{"5"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"0 {-} 5"}\NormalTok{, }\StringTok{"2.5 {-} 7.5"}\NormalTok{),}
               \StringTok{"10"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"5 {-} 10"}\NormalTok{, }\StringTok{"7.5 {-} 12.5"}\NormalTok{),}
               \StringTok{"15"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"10 {-} 15"}\NormalTok{, }\StringTok{"12.5 {-} 17.5"}\NormalTok{),}
               \StringTok{"20"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"15 {-} 20"}\NormalTok{, }\StringTok{"17.5 {-} 22.5"}\NormalTok{),}
               \StringTok{"25"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"20 {-} 25"}\NormalTok{, }\StringTok{"22.5 {-} 27.5"}\NormalTok{),}
               \StringTok{"30"} \OtherTok{=} \FunctionTok{c}\NormalTok{(}\StringTok{"25 {-} 30"}\NormalTok{, }\StringTok{"27.5 {-} 30"}\NormalTok{))}

\DocumentationTok{\#\# merge seagrass0tmp into seagrassHabitat to account for 0m samples,}
\DocumentationTok{\#\# set class for date, datetime variables}
\NormalTok{seagrassHabitatFull }\OtherTok{\textless{}{-}} 
  \FunctionTok{rbind}\NormalTok{(seagrass0tmp, seagrassHabitat) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(transect\_dist }\SpecialCharTok{!=} \StringTok{"0 {-} 2.5"}\NormalTok{)  }\SpecialCharTok{\%\textgreater{}\%} \CommentTok{\# already captured in seagrass0tmp }
  \FunctionTok{droplevels}\NormalTok{(.)  }\CommentTok{\# remove now unused factor levels}
\end{Highlighting}
\end{Shaded}

With the distances of habitat and density surveys now corresponding, we can
now merge these two datasets plus there coordinates together, combine
redundant fields, and remove unnecessary fields.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Merge seagrassHabitatFull with seagrassDensity, then coordinates}
\NormalTok{seagrass }\OtherTok{\textless{}{-}} 
  \FunctionTok{full\_join}\NormalTok{(seagrassHabitatFull, seagrassDensity, }
            \AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"organization"}\NormalTok{,}
                   \StringTok{"work\_area"}\NormalTok{,}
                   \StringTok{"project"}\NormalTok{,}
                   \StringTok{"survey"}\NormalTok{,}
                   \StringTok{"site\_id"}\NormalTok{, }
                   \StringTok{"date"}\NormalTok{,}
                   \StringTok{"transect\_dist"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \CommentTok{\# merge hakai\_id.x and hakai\_id.y into single variable field;}
  \CommentTok{\# use combination of date, site\_id, transect\_dist, and field uid (hakai\_id }
  \CommentTok{\# when present)}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{field\_uid =} \FunctionTok{ifelse}\NormalTok{(sample\_collected }\SpecialCharTok{==} \ConstantTok{TRUE}\NormalTok{, hakai\_id.x, }\StringTok{"NA"}\NormalTok{),}
         \AttributeTok{hakai\_id =} \FunctionTok{paste}\NormalTok{(date, }\StringTok{"HAKAI:CALVERT"}\NormalTok{, site\_id, transect\_dist, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \CommentTok{\# below, aggregate metadata that didn\textquotesingle{}t merge naturally (ie. due to minor }
         \CommentTok{\# differences in watch time or depth gauges)}
         \AttributeTok{dive\_supervisor =}\NormalTok{ dive\_supervisor.x,}
         \AttributeTok{collected\_start =} \FunctionTok{ymd\_hms}\NormalTok{(}\FunctionTok{ifelse}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(collected\_start.x),}
\NormalTok{                                          collected\_start.y, }
\NormalTok{                                          collected\_start.x)),}
         \AttributeTok{collected\_end   =} \FunctionTok{ymd\_hms}\NormalTok{(}\FunctionTok{ifelse}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(collected\_start.x),}
\NormalTok{                                          collected\_start.y,}
\NormalTok{                                          collected\_start.x)),}
         \AttributeTok{depth\_m         =} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(depth.x), depth.y, depth.x),}
         \AttributeTok{sampling\_bout   =}\NormalTok{ sampling\_bout.x) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{left\_join}\NormalTok{(., coordinates,  }\CommentTok{\# add coordinates}
            \AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"site\_id"} \OtherTok{=} \StringTok{"Point.Name"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{( }\SpecialCharTok{{-}} \FunctionTok{c}\NormalTok{(X.x, X.y, hakai\_id.x, hakai\_id.y,  }\CommentTok{\# remove unnecessary variables}
\NormalTok{              dive\_supervisor.x, dive\_supervisor.y,}
\NormalTok{              collected\_start.x, collected\_start.y,}
\NormalTok{              collected\_end.x, collected\_end.y,}
\NormalTok{              depth.x, depth.y,}
\NormalTok{              sampling\_bout.x, sampling\_bout.y)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{density\_msq =} \FunctionTok{as.character}\NormalTok{(density\_msq),}
         \AttributeTok{canopy\_height\_cm =} \FunctionTok{as.character}\NormalTok{(canopy\_height\_cm),}
         \AttributeTok{flowering\_shoots =} \FunctionTok{as.character}\NormalTok{(flowering\_shoots),}
         \AttributeTok{depth\_m =} \FunctionTok{as.character}\NormalTok{(depth\_m)) }\SpecialCharTok{\%T\textgreater{}\%}
  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 3,743
## Columns: 38
## $ organization     <chr> "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI", "HAKAI",~
## $ work_area        <chr> "CALVERT", "CALVERT", "CALVERT", "CALVERT", "CALVERT"~
## $ project          <chr> "MARINEGEO", "MARINEGEO", "MARINEGEO", "MARINEGEO", "~
## $ survey           <chr> "CHOKED_PASS", "CHOKED_PASS", "CHOKED_PASS", "PRUTH_B~
## $ site_id          <chr> "CHOKED_PASS_INTERIOR6", "CHOKED_PASS_EDGE1", "CHOKED~
## $ date             <date> 2017-11-22, 2017-05-19, 2017-05-19, 2017-07-03, 2017~
## $ collector.x      <chr> "zach", "kyle", NA, "tanya", "zach", "zach", "zach", ~
## $ sample_type.x    <chr> "seagrass_habitat", "seagrass_habitat", "seagrass_hab~
## $ transect_dist    <fct> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,~
## $ bag_uid          <chr> "10883", "7119", "7031", "2352", "10255", "10023", "1~
## $ bag_number       <chr> "3557", "800", "301", "324", "3506", "3555", "3534", ~
## $ density_range    <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ substrate        <chr> "sand,shell hash", "sand,shell hash", "sand,shell has~
## $ patchiness       <chr> "< 1", "< 1", "< 1", "< 1", "< 1", "05-Apr", "04-Mar"~
## $ adj_habitat_1    <chr> "seagrass", "sand", "standing kelp", "seagrass", "sea~
## $ adj_habitat_2    <chr> NA, NA, NA, NA, NA, NA, "standing kelp", NA, NA, NA, ~
## $ sample_collected <chr> "TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUE~
## $ vegetation_1     <chr> NA, "des", "des", "zm", "des", NA, NA, NA, NA, NA, NA~
## $ vegetation_2     <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, "~
## $ comments.x       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ quality_log.x    <chr> "1: Flowering shoots 0 for entire transects", NA, NA,~
## $ collector.y      <chr> "derek", "ondine", "ondine", "derek", "derek", "derek~
## $ sample_type.y    <chr> "seagrass_density", "seagrass_density", "seagrass_den~
## $ density          <dbl> 4, 10, 6, 13, 6, 1, 2, 6, 21, 3, 7, 4, 3, 14, 17, 11,~
## $ density_msq      <chr> "64", "160", "96", "208", "96", "16", "32", "96", "33~
## $ canopy_height_cm <chr> "80", "80", "110", "60", "125", "100", "100", "125", ~
## $ flowering_shoots <chr> "0", NA, NA, NA, NA, NA, NA, "0", NA, NA, NA, "0", NA~
## $ comments.y       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N~
## $ quality_log.y    <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, "~
## $ field_uid        <chr> "10883", "07119", "07031", "02352", "10255", "10023",~
## $ hakai_id         <chr> "2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0", "~
## $ dive_supervisor  <chr> "gillian", "gillian,gillian.sadlierbrown", "gillian,g~
## $ collected_start  <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ collected_end    <dttm> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ depth_m          <chr> "9.2", "3.4", "4.8", "2.4", "5.3", "5.6", "4.4", "2.5~
## $ sampling_bout    <chr> "6", "1", "3", "5", "5", "3", "5", "2", "1", "2", "6"~
## $ Decimal.Lat      <dbl> 51.67482, 51.67882, 51.67493, 51.64532, 51.67349, 51.~
## $ Decimal.Long     <dbl> -128.1195, -128.1148, -128.1237, -128.1193, -128.1180~
\end{verbatim}

\hypertarget{convert-data-to-darwin-core---extended-measurement-or-fact-format}{%
\subsection{Convert Data to Darwin Core - Extended Measurement or Fact format}\label{convert-data-to-darwin-core---extended-measurement-or-fact-format}}

The Darwin Core ExtendedMeasurementOrFact (eMoF) extension bases records
around a core event (rather than occurrence as in standard Darwin Core),
allowing for additional measurement variables to be associated with
occurrence data.

\hypertarget{add-event-id-and-occurrence-id-variables-to-dataset}{%
\subsubsection{Add Event ID and Occurrence ID variables to dataset}\label{add-event-id-and-occurrence-id-variables-to-dataset}}

As this dataset will be annually updated, rather than using
natural keys (ie. using package::uuid to autogenerate) for event and
occurence IDs, here we will use surrogate keys made up of a concatenation
of date survey, transect location, observation distance, and sample ID
(for occurrenceID, when a sample is present).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# create and populate eventID variable}
\DocumentationTok{\#\# currently only event is used, but additional surveys and abiotic data}
\DocumentationTok{\#\# are associated with parent events that may be included at a later date}
\NormalTok{seagrass}\SpecialCharTok{$}\NormalTok{eventID }\OtherTok{\textless{}{-}}\NormalTok{ seagrass}\SpecialCharTok{$}\NormalTok{hakai\_id}

\CommentTok{\# create and populate occurrenceID; combine eventID with transect\_dist }
\CommentTok{\# and field\_uid}
\DocumentationTok{\#\# in the event of \textless{}NA\textgreater{} field\_uid, no sample was collected, but}
\DocumentationTok{\#\# measurements and occurrence are still taken; no further subsamples}
\DocumentationTok{\#\# are associated with \textless{}NA\textgreater{} field\_uids}
\NormalTok{seagrass}\SpecialCharTok{$}\NormalTok{occurrenceID }\OtherTok{\textless{}{-}} 
  \FunctionTok{with}\NormalTok{(seagrass, }
       \FunctionTok{paste}\NormalTok{(eventID, transect\_dist, field\_uid, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{create-event-occurrence-and-emof-tables}{%
\subsubsection{Create Event, Occurrence, and eMoF tables}\label{create-event-occurrence-and-emof-tables}}

Now that we've created eventIDs and occurrenceIDs to connect all the
variables together, we can begin to create the Event, Occurrence,
and extended Measurement or Fact table necessary for DarwinCore
compliant datasets

\hypertarget{event-table}{%
\paragraph{Event Table}\label{event-table}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# subset seagrass to create event table}
\NormalTok{seagrassEvent }\OtherTok{\textless{}{-}}
\NormalTok{  seagrass }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  distinct }\SpecialCharTok{\%\textgreater{}\%}  \CommentTok{\# some duplicates in data stemming from database conflicts}
  \FunctionTok{select}\NormalTok{(date,}
\NormalTok{         Decimal.Lat, Decimal.Long, transect\_dist,}
\NormalTok{         depth\_m, eventID) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{eventDate                     =}\NormalTok{ date,}
         \AttributeTok{decimalLatitude               =}\NormalTok{ Decimal.Lat,}
         \AttributeTok{decimalLongitude              =}\NormalTok{ Decimal.Long,}
         \AttributeTok{coordinateUncertaintyInMeters =}\NormalTok{ transect\_dist,}
         \AttributeTok{minimumDepthInMeters          =}\NormalTok{ depth\_m,}
         \AttributeTok{maximumDepthInMeters          =}\NormalTok{ depth\_m) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{geodeticDatum  =} \StringTok{"WGS84"}\NormalTok{,}
         \AttributeTok{samplingEffort =} \StringTok{"30 metre transect"}\NormalTok{) }\SpecialCharTok{\%T\textgreater{}\%}\NormalTok{ glimpse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 3,659
## Columns: 8
## $ eventDate                     <date> 2017-11-22, 2017-05-19, 2017-05-19, 201~
## $ decimalLatitude               <dbl> 51.67482, 51.67882, 51.67493, 51.64532, ~
## $ decimalLongitude              <dbl> -128.1195, -128.1148, -128.1237, -128.11~
## $ coordinateUncertaintyInMeters <fct> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0~
## $ maximumDepthInMeters          <chr> "9.2", "3.4", "4.8", "2.4", "5.3", "5.6"~
## $ eventID                       <chr> "2017-11-22:HAKAI:CALVERT:CHOKED_PASS_IN~
## $ geodeticDatum                 <chr> "WGS84", "WGS84", "WGS84", "WGS84", "WGS~
## $ samplingEffort                <chr> "30 metre transect", "30 metre transect"~
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# save event table to csv}
\FunctionTok{write.csv}\NormalTok{(seagrassEvent, }\StringTok{"processed\_data/hakaiSeagrassDwcEvent.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{occurrence-table}{%
\paragraph{Occurrence Table}\label{occurrence-table}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# subset seagrass to create occurrence table}
\NormalTok{seagrassOccurrence }\OtherTok{\textless{}{-}}
\NormalTok{  seagrass }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  distinct }\SpecialCharTok{\%\textgreater{}\%}  \CommentTok{\# some duplicates in data stemming from database conflicts}
  \FunctionTok{select}\NormalTok{(eventID, occurrenceID) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{basisOfRecord =} \StringTok{"HumanObservation"}\NormalTok{,}
         \AttributeTok{scientificName   =} \StringTok{"Zostera subg. Zostera marina"}\NormalTok{,}
         \AttributeTok{occurrenceStatus =} \StringTok{"present"}\NormalTok{)}

\CommentTok{\# Taxonomic name matching}
\CommentTok{\# in addition to the above metadata, DarwinCore format requires further}
\CommentTok{\# taxonomic data that can be acquired through the WoRMS register.}
\DocumentationTok{\#\# Load taxonomic info, downloaded via WoRMS tool}
\CommentTok{\# zmWorms \textless{}{-} }
\CommentTok{\#   read.delim("raw\_data/zmworms\_matched.txt",}
\CommentTok{\#              header = TRUE,}
\CommentTok{\#              nrows  = 1)}

\NormalTok{zmWorms }\OtherTok{\textless{}{-}} \FunctionTok{wm\_record}\NormalTok{(}\AttributeTok{id =} \DecValTok{145795}\NormalTok{)}

\CommentTok{\# join WoRMS name with seagrassOccurrence create above}
\NormalTok{seagrassOccurrence }\OtherTok{\textless{}{-}} 
  \FunctionTok{full\_join}\NormalTok{(seagrassOccurrence, zmWorms, }
            \AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"scientificName"} \OtherTok{=} \StringTok{"scientificname"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(eventID, occurrenceID, basisOfRecord, scientificName, occurrenceStatus, AphiaID,}
\NormalTok{         url, authority, status, unacceptreason, taxonRankID, rank,}
\NormalTok{         valid\_AphiaID, valid\_name, valid\_authority, parentNameUsageID,}
\NormalTok{         kingdom, phylum, class, order, family, genus, citation, lsid,}
\NormalTok{         isMarine, match\_type, modified) }\SpecialCharTok{\%T\textgreater{}\%}
\NormalTok{  glimpse}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 3,659
## Columns: 27
## $ eventID           <chr> "2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0", ~
## $ occurrenceID      <chr> "2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0:0:~
## $ basisOfRecord     <chr> "HumanObservation", "HumanObservation", "HumanObserv~
## $ scientificName    <chr> "Zostera subg. Zostera marina", "Zostera subg. Zoste~
## $ occurrenceStatus  <chr> "present", "present", "present", "present", "present~
## $ AphiaID           <int> 145795, 145795, 145795, 145795, 145795, 145795, 1457~
## $ url               <chr> "https://www.marinespecies.org/aphia.php?p=taxdetail~
## $ authority         <chr> "Linnaeus, 1753", "Linnaeus, 1753", "Linnaeus, 1753"~
## $ status            <chr> "accepted", "accepted", "accepted", "accepted", "acc~
## $ unacceptreason    <lgl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, ~
## $ taxonRankID       <int> 220, 220, 220, 220, 220, 220, 220, 220, 220, 220, 22~
## $ rank              <chr> "Species", "Species", "Species", "Species", "Species~
## $ valid_AphiaID     <int> 145795, 145795, 145795, 145795, 145795, 145795, 1457~
## $ valid_name        <chr> "Zostera subg. Zostera marina", "Zostera subg. Zoste~
## $ valid_authority   <chr> "Linnaeus, 1753", "Linnaeus, 1753", "Linnaeus, 1753"~
## $ parentNameUsageID <int> 370435, 370435, 370435, 370435, 370435, 370435, 3704~
## $ kingdom           <chr> "Plantae", "Plantae", "Plantae", "Plantae", "Plantae~
## $ phylum            <chr> "Tracheophyta", "Tracheophyta", "Tracheophyta", "Tra~
## $ class             <chr> "Magnoliopsida", "Magnoliopsida", "Magnoliopsida", "~
## $ order             <chr> "Alismatales", "Alismatales", "Alismatales", "Alisma~
## $ family            <chr> "Zosteraceae", "Zosteraceae", "Zosteraceae", "Zoster~
## $ genus             <chr> "Zostera", "Zostera", "Zostera", "Zostera", "Zostera~
## $ citation          <chr> "WoRMS (2022). Zostera subg. Zostera marina Linnaeus~
## $ lsid              <chr> "urn:lsid:marinespecies.org:taxname:145795", "urn:ls~
## $ isMarine          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1~
## $ match_type        <chr> "exact", "exact", "exact", "exact", "exact", "exact"~
## $ modified          <chr> "2008-12-09T10:03:16.140Z", "2008-12-09T10:03:16.140~
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# save occurrence table to csv}
\FunctionTok{write.csv}\NormalTok{(seagrassOccurrence, }\StringTok{"processed\_data/hakaiSeagrassDwcOccurrence.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{extended-measurementorfact-table}{%
\paragraph{Extended MeasurementOrFact table}\label{extended-measurementorfact-table}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{seagrassMof }\OtherTok{\textless{}{-}}
\NormalTok{  seagrass }\SpecialCharTok{\%\textgreater{}\%}
  \CommentTok{\# select variables for eMoF table}
  \FunctionTok{select}\NormalTok{(date,}
\NormalTok{         eventID, survey, site\_id, transect\_dist,}
\NormalTok{         substrate, patchiness, adj\_habitat\_1, adj\_habitat\_2,}
\NormalTok{         vegetation\_1, vegetation\_2,}
\NormalTok{         density\_msq, canopy\_height\_cm, flowering\_shoots) }\SpecialCharTok{\%\textgreater{}\%}
  \CommentTok{\# split substrate into two variables (currently holds two substrate type in same variable)}
  \FunctionTok{separate}\NormalTok{(substrate, }\AttributeTok{sep =} \StringTok{","}\NormalTok{, }\AttributeTok{into =} \FunctionTok{c}\NormalTok{(}\StringTok{"substrate\_1"}\NormalTok{, }\StringTok{"substrate\_2"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \CommentTok{\# change variables names to match NERC database (or to be more descriptive where none exist)}
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{measurementDeterminedDate   =}\NormalTok{ date,}
         \AttributeTok{SubstrateTypeA              =}\NormalTok{ substrate\_1,}
         \AttributeTok{SubstrateTypeB              =}\NormalTok{ substrate\_2,}
         \AttributeTok{BarePatchLengthWithinSeagrass =}\NormalTok{ patchiness,}
         \AttributeTok{PrimaryAdjacentHabitat      =}\NormalTok{ adj\_habitat\_1,}
         \AttributeTok{SecondaryAdjacentHabitat    =}\NormalTok{ adj\_habitat\_2,}
         \AttributeTok{PrimaryAlgaeSp              =}\NormalTok{ vegetation\_1,}
         \AttributeTok{SecondaryAlgaeSp            =}\NormalTok{ vegetation\_2,}
         \AttributeTok{BedAbund                    =}\NormalTok{ density\_msq,}
         \AttributeTok{CanopyHeight                =}\NormalTok{ canopy\_height\_cm,}
         \AttributeTok{FloweringBedAbund           =}\NormalTok{ flowering\_shoots) }\SpecialCharTok{\%\textgreater{}\%}  
  \CommentTok{\# reformat variables into DwC MeasurementOrFact format}
  \CommentTok{\# (single values variable, with measurement type, unit, etc. variables)}
  \FunctionTok{pivot\_longer}\NormalTok{( }\SpecialCharTok{{-}} \FunctionTok{c}\NormalTok{(measurementDeterminedDate, eventID, survey, site\_id, transect\_dist),}
                \AttributeTok{names\_to =} \StringTok{"measurementType"}\NormalTok{,}
                \AttributeTok{values\_to =} \StringTok{"measurementValue"}\NormalTok{,}
                \AttributeTok{values\_ptypes =} \FunctionTok{list}\NormalTok{(}\AttributeTok{measurementValue =} \StringTok{"character"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \CommentTok{\# use measurement type to fill in remainder of variables relating to }
  \CommentTok{\# NERC vocabulary and metadata fields}
  \FunctionTok{mutate}\NormalTok{(}
    \AttributeTok{measurementTypeID =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL02/"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"CanopyHeight"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P01/current/OBSMAXLX/"}\NormalTok{,}
      \CommentTok{\# measurementType == "BarePatchWithinSeagrass" \textasciitilde{} "",}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"FloweringBedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL02/"}\NormalTok{),}
    \AttributeTok{measurementUnit =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"Number per square metre"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"CanopyHeight"} \SpecialCharTok{\textasciitilde{}} \StringTok{"Centimetres"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BarePatchhLengthWithinSeagrass"} \SpecialCharTok{\textasciitilde{}} \StringTok{"Metres"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"FloweringBedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"Number per square metre"}\NormalTok{),}
    \AttributeTok{measurementUnitID =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/UPMS/"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"CanopyHeight"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/ULCM/"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BarePatchhLengthWithinSeagrass"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/ULAA/2/"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"FloweringBedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/UPMS/"}\NormalTok{),}
    \AttributeTok{measurementAccuracy =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"CanopyHeight"} \SpecialCharTok{\textasciitilde{}} \DecValTok{5}\NormalTok{),}
    \AttributeTok{measurementMethod =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"25cmx25cm quadrat count"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"CanopyHeight"} \SpecialCharTok{\textasciitilde{}} \StringTok{"in situ with ruler"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"BarePatchhLengthWithinSeagrass"} \SpecialCharTok{\textasciitilde{}} \StringTok{"estimated along transect line"}\NormalTok{,}
\NormalTok{      measurementType }\SpecialCharTok{==} \StringTok{"FloweringBedAbund"} \SpecialCharTok{\textasciitilde{}} \StringTok{"25cmx25cm quadrat count"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(eventID, measurementDeterminedDate, measurementType, measurementValue,}
\NormalTok{         measurementTypeID, measurementUnit, measurementUnitID, measurementAccuracy,}
\NormalTok{         measurementMethod) }\SpecialCharTok{\%T\textgreater{}\%}
\CommentTok{\#  select(!c(survey, site\_id, transect\_dist)) \%T\textgreater{}\%}
  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Rows: 37,430
## Columns: 9
## $ eventID                   <chr> "2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERI~
## $ measurementDeterminedDate <date> 2017-11-22, 2017-11-22, 2017-11-22, 2017-11~
## $ measurementType           <chr> "SubstrateTypeA", "SubstrateTypeB", "BarePat~
## $ measurementValue          <chr> "sand", "shell hash", "< 1", "seagrass", NA,~
## $ measurementTypeID         <chr> NA, NA, NA, NA, NA, NA, NA, "http://vocab.ne~
## $ measurementUnit           <chr> NA, NA, NA, NA, NA, NA, NA, "Number per squa~
## $ measurementUnitID         <chr> NA, NA, NA, NA, NA, NA, NA, "http://vocab.ne~
## $ measurementAccuracy       <dbl> NA, NA, NA, NA, NA, NA, NA, NA, 5, NA, NA, N~
## $ measurementMethod         <chr> NA, NA, NA, NA, NA, NA, NA, "25cmx25cm quadr~
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# save eMoF table to csv}
\FunctionTok{write.csv}\NormalTok{(seagrassMof, }\StringTok{"processed\_data/hakaiSeagrassDwcEmof.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{session-info}{%
\subsection{Session Info}\label{session-info}}

Print session information below in case necessary for future reference

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Print Session Info for future reference}
\FunctionTok{sessionInfo}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## R version 4.1.1 (2021-08-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] worrms_0.4.2    magrittr_2.0.2  knitr_1.37      lubridate_1.8.0
##  [5] here_1.0.1      forcats_0.5.1   stringr_1.4.0   dplyr_1.0.8    
##  [9] purrr_0.3.4     readr_2.1.2     tidyr_1.2.0     tibble_3.1.6   
## [13] ggplot2_3.3.5   tidyverse_1.3.1
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.8       assertthat_0.2.1 rprojroot_2.0.2  digest_0.6.29   
##  [5] utf8_1.2.2       R6_2.5.1         cellranger_1.1.0 backports_1.4.1 
##  [9] reprex_2.0.1     evaluate_0.14    httr_1.4.2       pillar_1.7.0    
## [13] rlang_1.0.1      curl_4.3.2       readxl_1.3.1     rstudioapi_0.13 
## [17] jquerylib_0.1.4  rmarkdown_2.11   urltools_1.7.3   triebeard_0.3.0 
## [21] bit_4.0.4        munsell_0.5.0    broom_0.7.12     compiler_4.1.1  
## [25] modelr_0.1.8     xfun_0.29        pkgconfig_2.0.3  htmltools_0.5.2 
## [29] tidyselect_1.1.1 httpcode_0.3.0   bookdown_0.24    fansi_1.0.2     
## [33] crayon_1.4.2     tzdb_0.2.0       dbplyr_2.1.1     withr_2.4.3     
## [37] crul_1.2.0       grid_4.1.1       jsonlite_1.7.3   gtable_0.3.0    
## [41] lifecycle_1.0.1  DBI_1.1.2        scales_1.1.1     cli_3.1.1       
## [45] stringi_1.7.6    vroom_1.5.7      fs_1.5.2         xml2_1.3.3      
## [49] ellipsis_0.3.2   generics_0.1.2   vctrs_0.3.8      tools_4.1.1     
## [53] bit64_4.0.5      glue_1.6.1       hms_1.1.1        parallel_4.1.1  
## [57] fastmap_1.1.0    yaml_2.2.2       colorspace_2.0-2 rvest_1.0.2     
## [61] haven_2.4.3
\end{verbatim}

\hypertarget{trawl-data}{%
\section{Trawl Data}\label{trawl-data}}

One of the more common datasets that can be standardized to Darwin Core and integrated within OBIS is catch data from e.g.~a trawl sampling event, or a zooplankton net tow. Of special concern here are datasets that include both a total (species-specific) catch weight, in addition to individual measurements (for a subset of the overall data). In this case, through our standardization to Darwin Core, we want to ensure that data users understand that the individual measurements are a part of, or subset of, the overall (species-specific) record, whilst at the same time ensure that data providers are not duplicating occurrence records to OBIS.

The GitHub issue related to application is can be found \href{https://github.com/iobis/env-data/issues/10}{here}

\hypertarget{workflow-overview}{%
\section{Workflow Overview}\label{workflow-overview}}

In our current setup, this relationship between the overall catch data and subsetted information is provided in the \href{https://tools.gbif.org/dwca-validator/extension.do?id=dwc:ResourceRelationship}{resourceRelationship} extension. This extension \emph{cannot} currently be harvested by GBIF. The required terms for this extension are \texttt{resourceID}, \texttt{relatedResourceID}, \texttt{resourceRelationshipID} and \texttt{relationshipOfResource}. The \texttt{relatedResourceID} here refers to the \emph{object} of the relationship, whereas the \texttt{resourceID} refers to the \emph{subject} of the relationship:

\begin{itemize}
\tightlist
\item
  resourceRelationshipID: a unique identifier for the relationship between one resource (the subject) and another (relatedResource, object).
\item
  resourceID: a unique identifier for the resource that is the subject of the relationship.
\item
  relatedResourceID: a unique identifier for the resource that is the object of the relationship.
\item
  relationshipOfResource: The relationship of the subject (identified by the resourceID) to the object (relatedResourceID). The relationshipOfResource is a free text field.
\end{itemize}

A few resources have been published to OBIS that contain the resourceRelationship extension (examples). Here, I'll lay out the process and coding used for the \href{http://ipt.iobis.org/obiscanada/resource?r=trawl-catch-and-species-abundance-from-the-2019-gulf-of-alaska-international-year-of-the-salmon-expedition}{Trawl Catch and Species Abundance from the 2019 Gulf of Alaska International Year of the Salmon Expedition}. In the following code chunks some details are omitted to improve the readability - the overall code to standardize the catch data can be found \href{https://github.com/HakaiInstitute/iys-oos/blob/develop/datasets/GoA_2019/Trawl/2019/data_wrangle_trawl2019.Rmd}{here}. This dataset includes species-specific total catch data at multiple stations (sampling events). From each catch, individual measurements were also taken. Depending on the number of individual caught in the trawl, this was either the total number of species individuals caught, or only a subset (in case of large numbers of individuals caught).

In this specific data record, we created a single Event Core with three extensions: an \emph{occurrence} extension, \emph{measurement or fact} extension, and the \emph{resourceRelationship} extension. However, in this walk-through I'll only touch on the Event Core, occurrence extension and resourceRelationship extension.

The trawl data is part of a larger project collecting various data types related to salmon ocean ecology. Therefore, in our Event Core we nested information related to the sampling event in the specific layer. (include a visual representation of the schema). Prior to creating the Event Core, we ensured that e.g.~dates and times followed the correct ISO-8601 standards, and converted to the correct time zone.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Time is recorded numerically (1037 instead of 10:37), so need to change these columns:}
\NormalTok{trawl2019}\SpecialCharTok{$}\NormalTok{END\_DEPLOYMENT\_TIME }\OtherTok{\textless{}{-}} \FunctionTok{substr}\NormalTok{(}\FunctionTok{as.POSIXct}\NormalTok{(}\FunctionTok{sprintf}\NormalTok{(}\StringTok{"\%04.0f"}\NormalTok{, trawl2019}\SpecialCharTok{$}\NormalTok{END\_DEPLOYMENT\_TIME), }\AttributeTok{format =} \StringTok{"\%H\%M"}\NormalTok{), }\DecValTok{12}\NormalTok{, }\DecValTok{16}\NormalTok{)}
\NormalTok{trawl2019}\SpecialCharTok{$}\NormalTok{BEGIN\_RETRIEVAL\_TIME }\OtherTok{\textless{}{-}} \FunctionTok{substr}\NormalTok{(}\FunctionTok{as.POSIXct}\NormalTok{(}\FunctionTok{sprintf}\NormalTok{(}\StringTok{"\%04.0f"}\NormalTok{, trawl2019}\SpecialCharTok{$}\NormalTok{BEGIN\_RETRIEVAL\_TIME), }\AttributeTok{format =} \StringTok{"\%H\%M"}\NormalTok{), }\DecValTok{12}\NormalTok{, }\DecValTok{16}\NormalTok{)}
\CommentTok{\# Additionally, the vessel time is recorded in \textquotesingle{}Vladivostok\textquotesingle{} according to the metadata tab. This has to be converted to UTC.  }
\NormalTok{trawl2019 }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{eventDate\_start =} \FunctionTok{format\_iso\_8601}\NormalTok{(}\FunctionTok{as.POSIXct}\NormalTok{(}\FunctionTok{paste}\NormalTok{(EVENT\_DATE\_START, END\_DEPLOYMENT\_TIME),}
                                                      \AttributeTok{tz =} \StringTok{"Asia/Vladivostok"}\NormalTok{)),}
         \AttributeTok{eventDate\_start =} \FunctionTok{str\_replace}\NormalTok{(eventDate\_start, }\StringTok{"}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{+00:00"}\NormalTok{, }\StringTok{"Z"}\NormalTok{),}
         \AttributeTok{eventDate\_finish =} \FunctionTok{format\_iso\_8601}\NormalTok{(}\FunctionTok{as.POSIXct}\NormalTok{(}\FunctionTok{paste}\NormalTok{(EVENT\_DATE\_FINISH, BEGIN\_RETRIEVAL\_TIME),}
                                                       \AttributeTok{tz =} \StringTok{"Asia/Vladivostok"}\NormalTok{)),}
         \AttributeTok{eventDate\_finish =} \FunctionTok{str\_replace}\NormalTok{(eventDate\_finish, }\StringTok{"}\SpecialCharTok{\textbackslash{}\textbackslash{}}\StringTok{+00:00"}\NormalTok{, }\StringTok{"Z"}\NormalTok{),}
         \AttributeTok{eventDate =} \FunctionTok{paste}\NormalTok{(eventDate\_start, eventDate\_finish, }\AttributeTok{sep =} \StringTok{"/"}\NormalTok{),}
         \AttributeTok{project =} \StringTok{"IYS"}\NormalTok{,}
         \AttributeTok{cruise =} \FunctionTok{paste}\NormalTok{(project, }\StringTok{"GoA2019"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{), }
         \AttributeTok{station =} \FunctionTok{paste}\NormalTok{(cruise, TOW\_NUMBER, }\AttributeTok{sep=}\StringTok{":Stn"}\NormalTok{),}
         \AttributeTok{trawl =} \FunctionTok{paste}\NormalTok{(station, }\StringTok{"trawl"}\NormalTok{, }\AttributeTok{sep=}\StringTok{":"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Then we created the various layers of our Event Core. We created these layers/data frames from two separate datasets that data are pulled from - one dataset that contains the \emph{overall} catch data, and one dataset that contains the \emph{specimen} data:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{trawl2019\_allCatch }\OtherTok{\textless{}{-}} \FunctionTok{read\_excel}\NormalTok{(}\FunctionTok{here}\NormalTok{(}\StringTok{"Trawl"}\NormalTok{, }\StringTok{"2019"}\NormalTok{, }\StringTok{"raw\_data"}\NormalTok{, }
                                      \StringTok{"2019\_GoA\_Fish\_Trawl\_catchdata.xlsx"}\NormalTok{), }\AttributeTok{sheet =} \StringTok{"CATCH\_FINAL"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{project =} \StringTok{"IYS"}\NormalTok{,}
         \AttributeTok{cruise =} \FunctionTok{paste}\NormalTok{(project, }\StringTok{"GoA2019"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{station =} \FunctionTok{paste}\NormalTok{(cruise, }\StringTok{\textasciigrave{}}\AttributeTok{TOW\_NUMBER (number)}\StringTok{\textasciigrave{}}\NormalTok{, }\AttributeTok{sep =} \StringTok{":Stn"}\NormalTok{),}
         \AttributeTok{trawl =} \FunctionTok{paste}\NormalTok{(station, }\StringTok{"trawl"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{))}

\NormalTok{trawl2019\_specimen }\OtherTok{\textless{}{-}} \FunctionTok{read\_excel}\NormalTok{(}\FunctionTok{here}\NormalTok{(}\StringTok{"Trawl"}\NormalTok{, }\StringTok{"2019"}\NormalTok{, }\StringTok{"raw\_data"}\NormalTok{, }\StringTok{"2019\_GoA\_Fish\_Specimen\_data.xlsx"}\NormalTok{), }
                                 \AttributeTok{sheet =} \StringTok{"SPECIMEN\_FINAL"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{project =} \StringTok{"IYS"}\NormalTok{,}
         \AttributeTok{cruise =} \FunctionTok{paste}\NormalTok{(project, }\StringTok{"GoA2019"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{station =} \FunctionTok{paste}\NormalTok{(cruise, TOW\_NUMBER, }\AttributeTok{sep =} \StringTok{":Stn"}\NormalTok{),}
         \AttributeTok{trawl =} \FunctionTok{paste}\NormalTok{(station, }\StringTok{"trawl"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{sample =} \FunctionTok{paste}\NormalTok{(trawl, }\StringTok{"sample"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{sample =} \FunctionTok{paste}\NormalTok{(sample, }\FunctionTok{row\_number}\NormalTok{(), }\AttributeTok{sep =} \StringTok{""}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Next we created the Event Core, ensuring that we connect the data to the right layer (i.e.~date and time should be connected to the layer associated with the sampling event). Please note that because we are creating multiple layers and nesting information, and then at a later stage combining different tables, this results in cells being populated with \texttt{NA}. These have to be removed prior to publishing the Event Core through the IPT.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{trawl2019\_project }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ project) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{distinct}\NormalTok{(eventID) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{type =} \StringTok{"project"}\NormalTok{)}

\NormalTok{trawl2019\_cruise }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ cruise,}
         \AttributeTok{parentEventID =}\NormalTok{ project) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{distinct}\NormalTok{(eventID, }\AttributeTok{.keep\_all =} \ConstantTok{TRUE}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{type =} \StringTok{"cruise"}\NormalTok{)}

\NormalTok{trawl2019\_station }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ station,}
         \AttributeTok{parentEventID =}\NormalTok{ cruise) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{distinct}\NormalTok{(eventID, }\AttributeTok{.keep\_all =} \ConstantTok{TRUE}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{type =} \StringTok{"station"}\NormalTok{)}

\CommentTok{\# The coordinates associated to the trawl need to be presented in a LINESTRING.}
\CommentTok{\# END\_LONGITUDE\_DD needs to be inverted (has to be between {-}180 and 180, inclusive). }
\NormalTok{trawl2019\_coordinates }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ trawl,}
\NormalTok{         START\_LATITUDE\_DD,}
\NormalTok{         longitude,}
\NormalTok{         END\_LATITUDE\_DD,}
\NormalTok{         END\_LONGITUDE\_DD) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{END\_LONGITUDE\_DD =}\NormalTok{ END\_LONGITUDE\_DD }\SpecialCharTok{*} \SpecialCharTok{{-}}\DecValTok{1}\NormalTok{,}
         \AttributeTok{footprintWKT =} \FunctionTok{paste}\NormalTok{(}\StringTok{"LINESTRING ("}\NormalTok{, longitude, START\_LATITUDE\_DD, }\StringTok{","}\NormalTok{, }
\NormalTok{                              END\_LONGITUDE\_DD, END\_LATITUDE\_DD, }\StringTok{")"}\NormalTok{)) }
\NormalTok{trawl2019\_linestring }\OtherTok{\textless{}{-}}\NormalTok{ obistools}\SpecialCharTok{::}\FunctionTok{calculate\_centroid}\NormalTok{(trawl2019\_coordinates}\SpecialCharTok{$}\NormalTok{footprintWKT)}
\NormalTok{trawl2019\_linestring }\OtherTok{\textless{}{-}} \FunctionTok{cbind}\NormalTok{(trawl2019\_coordinates, trawl2019\_linestring) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(eventID, footprintWKT, decimalLatitude, decimalLongitude, coordinateUncertaintyInMeters)}

\NormalTok{trawl2019\_trawl }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019 }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ trawl,}
         \AttributeTok{parentEventID =}\NormalTok{ station,}
\NormalTok{         eventDate,}
\NormalTok{         year,}
\NormalTok{         month,}
\NormalTok{         day) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{minimumDepthInMeters =} \DecValTok{0}\NormalTok{, }\CommentTok{\# headrope was at the surface}
         \AttributeTok{maximumDepthInMeters =}\NormalTok{ trawl2019}\SpecialCharTok{$}\NormalTok{MOUTH\_OPENING\_HEIGHT,}
         \AttributeTok{samplingProtocol =} \StringTok{"midwater trawl"}\NormalTok{, }\CommentTok{\# when available add DOI to paper here}
         \AttributeTok{locality =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{           trawl2019}\SpecialCharTok{$}\NormalTok{EVENT\_SUB\_TYPE }\SpecialCharTok{==} \StringTok{"Can EEZ"} \SpecialCharTok{\textasciitilde{}} \StringTok{"Canadian EEZ"}\NormalTok{),}
         \AttributeTok{locationID =} \FunctionTok{case\_when}\NormalTok{(}
\NormalTok{           trawl2019}\SpecialCharTok{$}\NormalTok{EVENT\_SUB\_TYPE }\SpecialCharTok{==} \StringTok{"Can EEZ"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://marineregions.org/mrgid/8493"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{left\_join}\NormalTok{(trawl2019\_linestring, }\AttributeTok{by =} \StringTok{"eventID"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{distinct}\NormalTok{(eventID, }\AttributeTok{.keep\_all =} \ConstantTok{TRUE}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{mutate}\NormalTok{(}\AttributeTok{type =} \StringTok{"midwater trawl"}\NormalTok{)}
  
\NormalTok{trawl2019\_sample }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019\_specimen }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ sample,}
         \AttributeTok{parentEventID =}\NormalTok{ trawl) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{distinct}\NormalTok{(eventID, }\AttributeTok{.keep\_all =} \ConstantTok{TRUE}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{type =} \StringTok{"individual sample"}\NormalTok{)}

\NormalTok{trawl2019\_event }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(trawl2019\_project, }
\NormalTok{                             trawl2019\_cruise,}
\NormalTok{                             trawl2019\_station,}
\NormalTok{                             trawl2019\_trawl,}
\NormalTok{                             trawl2019\_sample) }

\CommentTok{\# Remove NAs from the Event Core:}
\NormalTok{trawl2019\_event }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(trawl2019\_event, as.character)}
\NormalTok{trawl2019\_event[}\FunctionTok{is.na}\NormalTok{(trawl2019\_event)] }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{trawl2019\_event }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(trawl2019\_event)}
\end{Highlighting}
\end{Shaded}

\textbf{TO DO: Add visual of e.g.~the top 10 rows of the Event Core.}

Now that we created the Event Core, we create the occurrence extension. To do this, we create two separate occurrence data tables: one that includes the occurrence data for the \emph{total} catch, and one data table for the \emph{specimen} data. Finally, the Occurrence extension is created by combining these two data frames. Personally, I prefer to re-order it so it makes visual sense to me (nest the specimen occurrence records under their respective overall catch data).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{trawl2019\_allCatch\_worms }\OtherTok{\textless{}{-}}\NormalTok{ worrms}\SpecialCharTok{::}\FunctionTok{wm\_records\_names}\NormalTok{(}\FunctionTok{unique}\NormalTok{(trawl2019\_allCatch}\SpecialCharTok{$}\NormalTok{scientificname))}
\NormalTok{trawl2019\_occ }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(trawl2019\_allCatch, trawl2019\_allCatch\_worms, }\AttributeTok{by =} \StringTok{"scientificname"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ trawl,}
         \AttributeTok{specificEpithet =}\NormalTok{ species,}
         \AttributeTok{scientificNameAuthorship =}\NormalTok{ authority,}
         \AttributeTok{taxonomicStatus =}\NormalTok{ status,}
         \AttributeTok{taxonRank =}\NormalTok{ rank,}
         \AttributeTok{scientificName =}\NormalTok{ scientificname,}
         \AttributeTok{scientificNameID =}\NormalTok{ lsid,}
         \AttributeTok{individualCount =} \StringTok{\textasciigrave{}}\AttributeTok{CATCH\_COUNT (pieces)(**includes Russian expansion for some species)}\StringTok{\textasciigrave{}}\NormalTok{,}
         \AttributeTok{occurrenceRemarks =}\NormalTok{ COMMENTS) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{occurrenceID =} \FunctionTok{paste}\NormalTok{(eventID, }\StringTok{"occ"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{occurrenceID =} \FunctionTok{paste}\NormalTok{(occurrenceID, }\FunctionTok{row\_number}\NormalTok{(), }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{occurrenceStatus =} \StringTok{"present"}\NormalTok{,}
         \AttributeTok{sex =} \StringTok{""}\NormalTok{)}

\NormalTok{trawl2019\_catch\_ind\_worms }\OtherTok{\textless{}{-}}\NormalTok{ worrms}\SpecialCharTok{::}\FunctionTok{wm\_records\_names}\NormalTok{(}\FunctionTok{unique}\NormalTok{(trawl2019\_catch\_ind}\SpecialCharTok{$}\NormalTok{scientificname)) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{bind\_rows}\NormalTok{()}
\NormalTok{trawl2019\_catch\_ind\_occ }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(trawl2019\_catch\_ind, trawl2019\_catch\_ind\_worms, }\AttributeTok{by =} \StringTok{"scientificname"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{scientificNameAuthorship =}\NormalTok{ authority,}
         \AttributeTok{taxonomicStatus =}\NormalTok{ status,}
         \AttributeTok{taxonRank =}\NormalTok{ rank,}
         \AttributeTok{scientificName =}\NormalTok{ scientificname,}
         \AttributeTok{scientificNameID =}\NormalTok{ lsid) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{occurrenceID =} \FunctionTok{paste}\NormalTok{(eventID, }\StringTok{"occ"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{occurrenceStatus =} \StringTok{"present"}\NormalTok{,}
         \AttributeTok{individualCount =} \DecValTok{1}\NormalTok{)}

\CommentTok{\# Combine the two occurrence data frames:}
\NormalTok{trawl2019\_occ\_ext }\OtherTok{\textless{}{-}}\NormalTok{ dplyr}\SpecialCharTok{::}\FunctionTok{bind\_rows}\NormalTok{(trawl2019\_occ\_fnl, trawl2019\_catch\_ind\_fnl)}

\CommentTok{\# To re{-}order the occurrenceID, use following code:}
\NormalTok{order }\OtherTok{\textless{}{-}}\NormalTok{ stringr}\SpecialCharTok{::}\FunctionTok{str\_sort}\NormalTok{(trawl2019\_occ\_ext}\SpecialCharTok{$}\NormalTok{occurrenceID, }\AttributeTok{numeric=}\ConstantTok{TRUE}\NormalTok{)}
\NormalTok{trawl2019\_occ\_ext }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019\_occ\_ext[}\FunctionTok{match}\NormalTok{(order, trawl2019\_occ\_ext}\SpecialCharTok{$}\NormalTok{occurrenceID),] }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{basisOfRecord =} \StringTok{"HumanObservation"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbf{TO DO: Add visual of e.g.~the top 10 rows of the Occurrence extension.}

\textbf{Please note} that in the \emph{overall} species-specific occurrence data frame, \emph{individualCount} was not included. This term should not be used for abundance studies, but to avoid confusion and the appearance that the specimen records are an additional observation on top of the overall catch record, the \emph{individualCount} term was left blank for the overall catch data.

A resource relationship extension is created to further highlight that the individual samples in the occurrence extension are part of a larger overall catch that was also listed in the occurrence extension. In this extension, we wanted to make sure to highlight that the \emph{specimen} occurrence records are \emph{a subset of} the \emph{overall} catch data through the field \texttt{relationshipOfResource1}. Each of these relationships gets a unique \texttt{resourceRelationshipID}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{trawl\_resourceRelationship }\OtherTok{\textless{}{-}}\NormalTok{ trawl2019\_occ\_ext }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(eventID, occurrenceID, scientificName) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{resourceID =} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\StringTok{"sample"}\NormalTok{, trawl2019\_occ\_ext}\SpecialCharTok{$}\NormalTok{occurrenceID), trawl2019\_occ\_ext}\SpecialCharTok{$}\NormalTok{occurrenceID, }\ConstantTok{NA}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{eventID =} \FunctionTok{gsub}\NormalTok{(}\StringTok{":sample.*"}\NormalTok{, }\StringTok{""}\NormalTok{, trawl2019\_occ\_ext}\SpecialCharTok{$}\NormalTok{eventID)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(eventID, scientificName) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{n}\NormalTok{() }\SpecialCharTok{!=} \DecValTok{1}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{ungroup}\NormalTok{()}

\NormalTok{trawl\_resourceRelationship }\OtherTok{\textless{}{-}}\NormalTok{ trawl\_resourceRelationship }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{relatedResourceID =} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\StringTok{"sample"}\NormalTok{, trawl\_resourceRelationship}\SpecialCharTok{$}\NormalTok{occurrenceID), }\ConstantTok{NA}\NormalTok{, trawl\_resourceRelationship}\SpecialCharTok{$}\NormalTok{occurrenceID)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{relationshipOfResource =} \FunctionTok{ifelse}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(resourceID), }\StringTok{"is a subset of"}\NormalTok{, }\ConstantTok{NA}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
\NormalTok{  dplyr}\SpecialCharTok{::}\FunctionTok{arrange}\NormalTok{(eventID, scientificName) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{fill}\NormalTok{(relatedResourceID) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(resourceID))}

\NormalTok{order }\OtherTok{\textless{}{-}}\NormalTok{ stringr}\SpecialCharTok{::}\FunctionTok{str\_sort}\NormalTok{(trawl\_resourceRelationship}\SpecialCharTok{$}\NormalTok{resourceID, }\AttributeTok{numeric =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{trawl\_resourceRelationship }\OtherTok{\textless{}{-}}\NormalTok{ trawl\_resourceRelationship[}\FunctionTok{match}\NormalTok{(order, trawl\_resourceRelationship}\SpecialCharTok{$}\NormalTok{resourceID),]}

\NormalTok{trawl\_resourceRelationship }\OtherTok{\textless{}{-}}\NormalTok{ trawl\_resourceRelationship }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{resourceRelationshipID =} \FunctionTok{paste}\NormalTok{(relatedResourceID, }\StringTok{"rr"}\NormalTok{, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{),}
         \AttributeTok{ID =} \FunctionTok{sprintf}\NormalTok{(}\StringTok{"\%03d"}\NormalTok{, }\FunctionTok{row\_number}\NormalTok{()),}
         \AttributeTok{resourceRelationshipID =} \FunctionTok{paste}\NormalTok{(resourceRelationshipID, ID, }\AttributeTok{sep =} \StringTok{":"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(eventID, resourceRelationshipID, resourceID, relationshipOfResource, relatedResourceID)}
\end{Highlighting}
\end{Shaded}

\textbf{TO DO: Add visual of e.g.~the top 10 rows of the ResourceRelationship extension.}

\hypertarget{faq}{%
\subsection{FAQ}\label{faq}}

\textbf{Q1}. Why not use the terms \emph{associatedOccurrence} or \emph{associatedTaxa}?
\textbf{A}. There seems to be a movement away from the term \emph{associatedOccurrence} as the \texttt{resourceRelationship} extension has a much broader use case. Some issues that were raised on GitHub exemplify this, see e.g.~\href{https://github.com/tdwg/dwc/issues/324}{here}. \emph{associatedTaxa} is used to provide identifiers or names of taxa and the associations of an Occurrence with them. This term is not apt for establishing relationships between taxa, only between specific Occurrences of an organism with other taxa. As noted on the \href{https://dwc.tdwg.org/terms/\#dwc:associatedTaxa}{TDWG website}, \emph{{[}\ldots{]} Note that the ResourceRelationship class is an alternative means of representing associations, and with more detail.} See also e.g.~\href{https://github.com/tdwg/dwc/issues/331}{this issue}.

\hypertarget{aligning-data-to-darwin-core---sampling-event-with-measurement-or-fact}{%
\section{Aligning Data to Darwin Core - Sampling Event with Measurement or Fact}\label{aligning-data-to-darwin-core---sampling-event-with-measurement-or-fact}}

Abby Benson\\
October 8,2019

\hypertarget{general-information-about-this-notebook}{%
\subsection{General information about this notebook}\label{general-information-about-this-notebook}}

This notebook was created for the IOOS DMAC Code Sprint Biological Data Session
The data in this notebook were created specifically as an example and meant solely to be
illustrative of the process for aligning data to the biological data standard - Darwin Core.
These data should not be considered actually occurrences of species and any measurements
are also contrived. This notebook is meant to provide a step by step process for taking
original data and aligning it to Darwin Core.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(readr)}
\FunctionTok{library}\NormalTok{(uuid)}
\FunctionTok{library}\NormalTok{(dplyr)}

\NormalTok{MadeUpDataForBiologicalDataTraining }\OtherTok{\textless{}{-}} \FunctionTok{read\_csv}\NormalTok{(}\StringTok{"\textasciitilde{}/OBIS/Reference Documentation/Presentations/IOOS DMAC Code Sprint/MadeUpDataForBiologicalDataTraining.csv"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Parsed with column specification:
## cols(
##   date = col_character(),
##   lat = col_double(),
##   lon = col_double(),
##   region = col_character(),
##   station = col_double(),
##   transect = col_double(),
##   `scientific name` = col_character(),
##   `percent cover` = col_double(),
##   depth = col_double(),
##   `bottom type` = col_character(),
##   rugosity = col_double(),
##   temperature = col_double()
## )
\end{verbatim}

First we need to to decide if we will provide an occurrence only version of the data or
a sampling event with measurement or facts version of the data. Occurrence only is easier
to create. It's only one file to produce. However, several pieces of information will be
left out if we choose that option. If we choose to do sampling event with measurement or
fact we'll be able to capture all of the data in the file creating a lossless version.
Here we decide to use the sampling event option to include as much information as we can.
First let's create the eventID and occurrenceID in the original file so that information
can be reused for all necessary files down the line.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{eventID }\OtherTok{\textless{}{-}} \FunctionTok{paste}\NormalTok{(MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{region, }
\NormalTok{                                                     MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{station, }
\NormalTok{                                                     MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{transect, }\AttributeTok{sep =} \StringTok{"\_"}\NormalTok{) }
\NormalTok{MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{occurrenceID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{occurrenceID }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(MadeUpDataForBiologicalDataTraining}\SpecialCharTok{$}\NormalTok{occurrenceID, }
                                       \ControlFlowTok{function}\NormalTok{(x) }\FunctionTok{UUIDgenerate}\NormalTok{(}\AttributeTok{use.time =} \ConstantTok{TRUE}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

We will need to create three separate files to comply with the sampling event format.
We'll start with the event file but we only need to include the columns that are relevant
to the event file.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event }\OtherTok{\textless{}{-}}\NormalTok{ MadeUpDataForBiologicalDataTraining[}\FunctionTok{c}\NormalTok{(}\StringTok{"date"}\NormalTok{, }\StringTok{"lat"}\NormalTok{, }\StringTok{"lon"}\NormalTok{, }\StringTok{"region"}\NormalTok{, }\StringTok{"station"}\NormalTok{,}
                                               \StringTok{"transect"}\NormalTok{, }\StringTok{"depth"}\NormalTok{, }\StringTok{"bottom type"}\NormalTok{, }\StringTok{"eventID"}\NormalTok{)]}
\end{Highlighting}
\end{Shaded}

Next we need to rename any columns of data that match directly to Darwin Core. We know
this based on our crosswalk spreadsheet CrosswalkToDarwinCore.csv

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event}\SpecialCharTok{$}\NormalTok{decimalLatitude }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\NormalTok{lat}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{decimalLongitude }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\NormalTok{lon}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{minimumDepthInMeters }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\NormalTok{depth}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{maximumDepthInMeters }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\NormalTok{depth}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{habitat }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{bottom type}\StringTok{\textasciigrave{}}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{island }\OtherTok{\textless{}{-}}\NormalTok{ event}\SpecialCharTok{$}\NormalTok{region}
\end{Highlighting}
\end{Shaded}

Let's see how it looks:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(event, }\AttributeTok{n =} \DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 10 x 15
##    date    lat   lon region station transect depth `bottom type` eventID
##    <chr> <dbl> <dbl> <chr>    <dbl>    <dbl> <dbl> <chr>         <chr>  
##  1 7/16~  18.3 -64.8 St. J~     250        1    25 shallow reef~ St. Jo~
##  2 7/16~  18.3 -64.8 St. J~     250        1    25 shallow reef~ St. Jo~
##  3 7/16~  18.3 -64.8 St. J~     250        1    25 shallow reef~ St. Jo~
##  4 7/16~  18.3 -64.8 St. J~     250        1    25 shallow reef~ St. Jo~
##  5 7/16~  18.3 -64.8 St. J~     250        2    35 complex back~ St. Jo~
##  6 7/16~  18.3 -64.8 St. J~     250        2    35 complex back~ St. Jo~
##  7 7/16~  18.3 -64.8 St. J~     250        2    35 complex back~ St. Jo~
##  8 7/16~  18.3 -64.8 St. J~     250        2    35 complex back~ St. Jo~
##  9 7/16~  18.3 -64.8 St. J~     250        3    85 deep reef     St. Jo~
## 10 7/16~  18.3 -64.8 St. J~     250        3    85 deep reef     St. Jo~
## # ... with 6 more variables: decimalLatitude <dbl>,
## #   decimalLongitude <dbl>, minimumDepthInMeters <dbl>,
## #   maximumDepthInMeters <dbl>, habitat <chr>, island <chr>
\end{verbatim}

We need to convert the date to ISO format

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event}\SpecialCharTok{$}\NormalTok{eventDate }\OtherTok{\textless{}{-}} \FunctionTok{as.Date}\NormalTok{(event}\SpecialCharTok{$}\NormalTok{date, }\AttributeTok{format =} \StringTok{"\%m/\%d/\%Y"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We will also have to add any missing required fields

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event}\SpecialCharTok{$}\NormalTok{basisOfRecord }\OtherTok{\textless{}{-}} \StringTok{"HumanObservation"}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{geodeticDatum }\OtherTok{\textless{}{-}} \StringTok{"EPSG:4326 WGS84"}
\end{Highlighting}
\end{Shaded}

Then we'll remove any columns that we no longer need to clean things up a bit.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event}\SpecialCharTok{$}\NormalTok{date }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{lat }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{lon }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{region }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{station }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{transect }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\NormalTok{depth }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{event}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{bottom type}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \ConstantTok{NULL}
\end{Highlighting}
\end{Shaded}

We have too many repeating rows of information. We can pare this down using eventID which
is a unique identifier for each sampling event in the data- which is six, three transects
per site.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{event }\OtherTok{\textless{}{-}}\NormalTok{ event[}\FunctionTok{which}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{duplicated}\NormalTok{(event}\SpecialCharTok{$}\NormalTok{eventID)),]}

\FunctionTok{head}\NormalTok{(event, }\AttributeTok{n =} \DecValTok{6}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 6 x 10
##   eventID decimalLatitude decimalLongitude minimumDepthInM~
##   <chr>             <dbl>            <dbl>            <dbl>
## 1 St. Jo~            18.3            -64.8               25
## 2 St. Jo~            18.3            -64.8               35
## 3 St. Jo~            18.3            -64.8               85
## 4 St. Jo~            18.3            -64.8               28
## 5 St. Jo~            18.3            -64.8               16
## 6 St. Jo~            18.3            -64.8               90
## # ... with 6 more variables: maximumDepthInMeters <dbl>, habitat <chr>,
## #   island <chr>, eventDate <date>, basisOfRecord <chr>,
## #   geodeticDatum <chr>
\end{verbatim}

Finally we write out the event file

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{write.csv}\NormalTok{(event, }\AttributeTok{file=}\StringTok{"MadeUpData\_event.csv"}\NormalTok{, }\AttributeTok{row.names=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{fileEncoding=}\StringTok{"UTF{-}8"}\NormalTok{, }\AttributeTok{quote=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Next we need to create the occurrence file. We start by creating the dataframe.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{occurrence }\OtherTok{\textless{}{-}}\NormalTok{ MadeUpDataForBiologicalDataTraining[}\FunctionTok{c}\NormalTok{(}\StringTok{"scientific name"}\NormalTok{, }\StringTok{"eventID"}\NormalTok{, }\StringTok{"occurrenceID"}\NormalTok{, }\StringTok{"percent cover"}\NormalTok{)] }
\end{Highlighting}
\end{Shaded}

Then we'll rename the columns that align directly with Darwin Core.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{occurrence}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}}\NormalTok{ occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{scientific name}\StringTok{\textasciigrave{}}
\end{Highlighting}
\end{Shaded}

Finally we'll add required information that's missing.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{occurrence}\SpecialCharTok{$}\NormalTok{occurrenceStatus }\OtherTok{\textless{}{-}}  \FunctionTok{ifelse}\NormalTok{ (occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{percent cover}\StringTok{\textasciigrave{}} \SpecialCharTok{==} \DecValTok{0}\NormalTok{, }\StringTok{"absent"}\NormalTok{, }\StringTok{"present"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{taxonomic-name-matching}{%
\subsection{Taxonomic Name Matching}\label{taxonomic-name-matching}}

A requirement for OBIS is that all scientific names match to the World Register of
Marine Species (WoRMS) and a scientificNameID is included. A scientificNameID looks
like this ``\url{urn:lsid:marinespecies.org:taxname:275730}'' with the last digits after
the colon being the WoRMS aphia ID. We'll need to go out to WoRMS to grab this
information.
Create a lookup table of unique scientific names

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lut\_worms }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{unique}\NormalTok{(occurrence\_only}\SpecialCharTok{$}\NormalTok{scientificName))}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(lut\_worms}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{unique(occurrence\_only$scientificName)}\StringTok{\textasciigrave{}}\NormalTok{)}
\NormalTok{lut\_worms}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{unique(occurrence\_only$scientificName)}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{as.character}\NormalTok{(lut\_worms}\SpecialCharTok{$}\NormalTok{scientificName)}
\end{Highlighting}
\end{Shaded}

Add the columns that we can grab information from WoRMS including the required scientificNameID.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{acceptedname }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{acceptedID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{scientificNameID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{kingdom }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{phylum }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{class }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{order }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{family }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{genus }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{scientificNameAuthorship }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{lut\_worms}\SpecialCharTok{$}\NormalTok{taxonRank }\OtherTok{\textless{}{-}} \StringTok{""}
\end{Highlighting}
\end{Shaded}

Taxonomic lookup using the library taxizesoap

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{nrow}\NormalTok{(lut\_worms))\{}
\NormalTok{  df }\OtherTok{\textless{}{-}} \FunctionTok{worms\_records}\NormalTok{(}\AttributeTok{scientific =}\NormalTok{ lut\_worms}\SpecialCharTok{$}\NormalTok{scientificName[i])}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{scientificNameID }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{lsid[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{acceptedname }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{valid\_name[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{acceptedID }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{valid\_AphiaID[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{kingdom }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{kingdom[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{phylum }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{phylum[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{class }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{class[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{order }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{order[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{family }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{family[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{genus }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{genus[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{scientificNameAuthorship }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{authority[}\DecValTok{1}\NormalTok{]}
\NormalTok{  lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{taxonRank }\OtherTok{\textless{}{-}}\NormalTok{ df}\SpecialCharTok{$}\NormalTok{rank[}\DecValTok{1}\NormalTok{]}
  \FunctionTok{message}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"Looking up information for species:"}\NormalTok{, lut\_worms[i,]}\SpecialCharTok{$}\NormalTok{scientificName))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Looking up information for species: Acropora cervicornis
\end{verbatim}

\begin{verbatim}
## Looking up information for species: Madracis auretenra
\end{verbatim}

\begin{verbatim}
## Looking up information for species: Mussa angulosa
\end{verbatim}

\begin{verbatim}
## Looking up information for species: Siderastrea radians
\end{verbatim}

Merge the lookup table of unique scientific names back with the occurrence data.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{merge}\NormalTok{(occurrence, lut\_worms, }\AttributeTok{by =} \StringTok{"scientificName"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We're going to remove any unnecessary columns to clean up the file

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{scientific name}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{percent cover}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \ConstantTok{NULL}
\end{Highlighting}
\end{Shaded}

Quick look at what we have before we write out the file

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(occurrence, }\AttributeTok{n =} \DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          scientificName        eventID
## 1  Acropora cervicornis St. John_250_1
## 2  Acropora cervicornis St. John_250_2
## 3  Acropora cervicornis St. John_250_3
## 4  Acropora cervicornis St. John_356_1
## 5  Acropora cervicornis St. John_356_2
## 6  Acropora cervicornis St. John_356_3
## 7    Madracis auretenra St. John_250_1
## 8    Madracis auretenra St. John_250_2
## 9    Madracis auretenra St. John_250_3
## 10   Madracis auretenra St. John_356_1
##                            occurrenceID occurrenceStatus
## 1  63be8f7e-ea9a-11e9-8649-49c6324f3a06           absent
## 2  63beb687-ea9a-11e9-8649-49c6324f3a06           absent
## 3  63beb68b-ea9a-11e9-8649-49c6324f3a06          present
## 4  63bedd76-ea9a-11e9-8649-49c6324f3a06          present
## 5  63bedd7a-ea9a-11e9-8649-49c6324f3a06          present
## 6  63bedd7e-ea9a-11e9-8649-49c6324f3a06          present
## 7  63beb684-ea9a-11e9-8649-49c6324f3a06          present
## 8  63beb688-ea9a-11e9-8649-49c6324f3a06          present
## 9  63beb68c-ea9a-11e9-8649-49c6324f3a06           absent
## 10 63bedd77-ea9a-11e9-8649-49c6324f3a06          present
##            acceptedname acceptedID
## 1  Acropora cervicornis     206989
## 2  Acropora cervicornis     206989
## 3  Acropora cervicornis     206989
## 4  Acropora cervicornis     206989
## 5  Acropora cervicornis     206989
## 6  Acropora cervicornis     206989
## 7    Madracis auretenra     430664
## 8    Madracis auretenra     430664
## 9    Madracis auretenra     430664
## 10   Madracis auretenra     430664
##                             scientificNameID  kingdom   phylum    class
## 1  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 2  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 3  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 4  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 5  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 6  urn:lsid:marinespecies.org:taxname:206989 Animalia Cnidaria Anthozoa
## 7  urn:lsid:marinespecies.org:taxname:430664 Animalia Cnidaria Anthozoa
## 8  urn:lsid:marinespecies.org:taxname:430664 Animalia Cnidaria Anthozoa
## 9  urn:lsid:marinespecies.org:taxname:430664 Animalia Cnidaria Anthozoa
## 10 urn:lsid:marinespecies.org:taxname:430664 Animalia Cnidaria Anthozoa
##           order         family    genus   scientificNameAuthorship
## 1  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 2  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 3  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 4  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 5  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 6  Scleractinia    Acroporidae Acropora            (Lamarck, 1816)
## 7  Scleractinia Pocilloporidae Madracis Locke, Weil & Coates, 2007
## 8  Scleractinia Pocilloporidae Madracis Locke, Weil & Coates, 2007
## 9  Scleractinia Pocilloporidae Madracis Locke, Weil & Coates, 2007
## 10 Scleractinia Pocilloporidae Madracis Locke, Weil & Coates, 2007
##    taxonRank
## 1    Species
## 2    Species
## 3    Species
## 4    Species
## 5    Species
## 6    Species
## 7    Species
## 8    Species
## 9    Species
## 10   Species
\end{verbatim}

Write out the file. All done with occurrence!

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{write.csv}\NormalTok{(occurrence, }\AttributeTok{file=}\StringTok{"MadeUpData\_Occurrence.csv"}\NormalTok{, }\AttributeTok{row.names=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{fileEncoding=}\StringTok{"UTF{-}8"}\NormalTok{, }\AttributeTok{quote=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The last file we need to create is the measurement or fact file. For this we need to
combine all of the measurements or facts that we want to include making sure to include
IDs from the BODC NERC vocabulary where possible.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{temperature }\OtherTok{\textless{}{-}}\NormalTok{ MadeUpDataForBiologicalDataTraining[}\FunctionTok{c}\NormalTok{(}\StringTok{"eventID"}\NormalTok{, }\StringTok{"temperature"}\NormalTok{, }\StringTok{"date"}\NormalTok{)]}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{occurrenceID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementType }\OtherTok{\textless{}{-}} \StringTok{"temperature"}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementTypeID }\OtherTok{\textless{}{-}} \StringTok{"http://vocab.nerc.ac.uk/collection/P25/current/WTEMP/"}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementValue }\OtherTok{\textless{}{-}}\NormalTok{ temperature}\SpecialCharTok{$}\NormalTok{temperature}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementUnit }\OtherTok{\textless{}{-}} \StringTok{"Celsius"}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementUnitID }\OtherTok{\textless{}{-}} \StringTok{"http://vocab.nerc.ac.uk/collection/P06/current/UPAA/"}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementAccuracy }\OtherTok{\textless{}{-}} \DecValTok{3}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementDeterminedDate }\OtherTok{\textless{}{-}} \FunctionTok{as.Date}\NormalTok{(temperature}\SpecialCharTok{$}\NormalTok{date, }\AttributeTok{format =} \StringTok{"\%m/\%d/\%Y"}\NormalTok{)}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{measurementMethod }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{temperature }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{temperature}\SpecialCharTok{$}\NormalTok{date }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{rugosity }\OtherTok{\textless{}{-}}\NormalTok{ MadeUpDataForBiologicalDataTraining[}\FunctionTok{c}\NormalTok{(}\StringTok{"eventID"}\NormalTok{, }\StringTok{"rugosity"}\NormalTok{, }\StringTok{"date"}\NormalTok{)]}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{occurrenceID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementType }\OtherTok{\textless{}{-}} \StringTok{"rugosity"}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementTypeID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementValue }\OtherTok{\textless{}{-}}\NormalTok{ rugosity}\SpecialCharTok{$}\NormalTok{rugosity}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementUnit }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementUnitID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measuremntAccuracy }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementDeterminedDate }\OtherTok{\textless{}{-}} \FunctionTok{as.Date}\NormalTok{(rugosity}\SpecialCharTok{$}\NormalTok{date, }\AttributeTok{format =} \StringTok{"\%m/\%d/\%Y"}\NormalTok{)}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{measurementMethod }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{rugosity }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{rugosity}\SpecialCharTok{$}\NormalTok{date }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{percentcover }\OtherTok{\textless{}{-}}\NormalTok{ MadeUpDataForBiologicalDataTraining[}\FunctionTok{c}\NormalTok{(}\StringTok{"eventID"}\NormalTok{, }\StringTok{"occurrenceID"}\NormalTok{, }\StringTok{"percent cover"}\NormalTok{, }\StringTok{"date"}\NormalTok{)]}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementType }\OtherTok{\textless{}{-}} \StringTok{"Percent Cover"}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementTypeID }\OtherTok{\textless{}{-}} \StringTok{"http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL10/"}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementValue }\OtherTok{\textless{}{-}}\NormalTok{ percentcover}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{percent cover}\StringTok{\textasciigrave{}}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementUnit }\OtherTok{\textless{}{-}} \StringTok{"Percent/100m\^{}2"}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementUnitID }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measuremntAccuracy }\OtherTok{\textless{}{-}} \DecValTok{5}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementDeterminedDate }\OtherTok{\textless{}{-}} \FunctionTok{as.Date}\NormalTok{(percentcover}\SpecialCharTok{$}\NormalTok{date, }\AttributeTok{format =} \StringTok{"\%m/\%d/\%Y"}\NormalTok{)}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{measurementMethod }\OtherTok{\textless{}{-}} \StringTok{""}
\NormalTok{percentcover}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{percent cover}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{percentcover}\SpecialCharTok{$}\NormalTok{date }\OtherTok{\textless{}{-}} \ConstantTok{NULL}
\NormalTok{measurementOrFact }\OtherTok{\textless{}{-}} \FunctionTok{rbind}\NormalTok{(temperature, rugosity, percentcover)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Error in match.names(clabs, names(xi)): names do not match previous names
\end{verbatim}

Let's check to see what it looks like

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{(measurementOrFact, }\AttributeTok{n =} \DecValTok{50}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 50 x 9
##    eventID occurrenceID measurementType measurementType~ measurementValue
##    <chr>   <chr>        <chr>           <chr>                       <dbl>
##  1 St. Jo~ ""           temperature     ""                           25.2
##  2 St. Jo~ ""           temperature     ""                           25.2
##  3 St. Jo~ ""           temperature     ""                           25.2
##  4 St. Jo~ ""           temperature     ""                           25.2
##  5 St. Jo~ ""           temperature     ""                           24.8
##  6 St. Jo~ ""           temperature     ""                           24.8
##  7 St. Jo~ ""           temperature     ""                           24.8
##  8 St. Jo~ ""           temperature     ""                           24.8
##  9 St. Jo~ ""           temperature     ""                           23.1
## 10 St. Jo~ ""           temperature     ""                           23.1
## # ... with 40 more rows, and 4 more variables: measurementUnit <chr>,
## #   measurementUnitID <chr>, measurementDeterminedDate <date>,
## #   measurementMethod <chr>
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{write.csv}\NormalTok{(measurementOrFact, }\AttributeTok{file=}\StringTok{"MadeUpData\_mof.csv"}\NormalTok{, }\AttributeTok{row.names=}\ConstantTok{FALSE}\NormalTok{, }\AttributeTok{fileEncoding=}\StringTok{"UTF{-}8"}\NormalTok{, }\AttributeTok{quote=}\ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{dealing-with-errors}{%
\chapter{Dealing with errors}\label{dealing-with-errors}}

Datasets can have a wide variety of errors that pop up during the darwin core alignment process. This chapter details
ways in which a data manager can identify, discuss, and resolve potential errors in the data.

It should be noted that, in most cases, the data manager/scientist aligning the data to darwin core should reach out
to the data originator to ensure the actions taken are not incorrectly representing the observations.

\hypertarget{example-using-github-to-resolve-errors}{%
\section{Example using GitHub to resolve errors}\label{example-using-github-to-resolve-errors}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Dataset sent to OBIS-USA via email.
\item
  OBIS-USA uploaded to IPT.
\item
  Once the data were uploaded, the IPT identified there was an issue with the \texttt{occurrenceID} field. The issue was then
  presented and discussed in a GitHub ticket:
  \href{https://github.com/ioos/bio_data_guide/issues/78}{\includegraphics{./figs/issue_78.png}}
\item
  The data manager uploaded the raw data and code to GitHub through the pull request below. This included a fix for
  the \texttt{occurrenceID} issue.
  \href{https://github.com/ioos/bio_data_guide/pull/77}{\includegraphics{./figs/PR_77.png}}
\item
  The OBIS node manager was notified of the availability of a revised dataset by pointing directly to the appropriate
  commit in GitHub:
  \href{https://github.com/ioos/bio_data_guide/commit/ef17f89f23316a6227fdf58de509582bd9854a55}{\includegraphics{./figs/commit_ef17f89.png}}
\item
  The OBIS node manager downloaded the data from the commit above and uploaded them to the IPT.
\item
  The IPT returned a summary of the dataset including that 434 records had invalid \texttt{scientificNameID} records in the
  occurrence file.
\item
  After some data sleuthing, the data manager noticed that the code accidentally removed trailing zeros from
  \texttt{scientificNameID} that ended in \texttt{0}:
  \href{https://nbviewer.org/github/ioos/bio_data_guide/blob/ef17f89f23316a6227fdf58de509582bd9854a55/datasets/AMBON_zooplankton/2017zooplanton_to_dwc\%282\%29_mmb.ipynb\#END-of-ID-generation}{\includegraphics{./figs/code_snippet.png}}
\item
  So, the data manager updated the code to resolve the issue and generate a new occurrence file.
  \href{https://github.com/ioos/bio_data_guide/pull/82}{\includegraphics{./figs/PR_82.png}}

  \begin{enumerate}
  \def\labelenumii{\arabic{enumii}.}
  \tightlist
  \item
    Here is fixing the \texttt{scientificNameID} generation: \includegraphics{./figs/code_change1.png}
  \item
    Here is removing the problematic code: \includegraphics{./figs/code_change2.png} \includegraphics{./figs/code_change3.png}
  \end{enumerate}
\item
  The revised occurrence file was then resubmitted to the OBIS node manager by pointing them at the appropriate commit
  record:
  \href{https://github.com/ioos/bio_data_guide/commit/a0919e5b788b0737fc4c8e2c4b874c2e287769dd}{\includegraphics{./figs/commit_a0919e.png}}
\item
  The OBIS node manager downloaded the data from the commit above and uploaded them to the IPT.
\item
  The IPT and OBIS landing page now indicated that no more issues with these data are present:
  \href{https://obis.org/dataset/bc01451e-d990-4ad1-8315-e3fb6e9cf461}{\includegraphics{./figs/obis_dataset.png}}
\end{enumerate}

\hypertarget{final-words}{%
\chapter{Final Words}\label{final-words}}

We have finished a nice book.

\hypertarget{tools}{%
\chapter{Tools}\label{tools}}

Below are some of the tools and packages used in workflows. R and Python package ``Type'' is BIO for packages specifically for biological applications, and GEN for generic packages.

\hypertarget{r}{%
\section{R}\label{r}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.28}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.39}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.33}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Type
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://bdverse.org/}{bdveRse} & BIO & A family of R packages for biodiversity data. \\
\href{https://cran.r-project.org/web/packages/ecocomDP/index.html}{ecocomDP} & BIO & Work with the Ecological Community Data Design Pattern. `ecocomDP' is a flexible data model for harmonizing ecological community surveys, in a research question agnostic format, from source data published across repositories, and with methods that keep the derived data up-to-date as the underlying sources change. \\
\href{https://ediorg.github.io/EMLassemblyline/}{EDIorg/EMLasseblyline} & BIO & For scientists and data managers to create high quality EML metadata for dataset publication. \\
\href{https://cran.r-project.org/web/packages/finch/index.html}{finch} & BIO & Parse Darwin Core Files \\
\href{https://iobis.github.io/obistools/}{iobis/obistools} & BIO & Tools for data enhancement and quality control. \\
\href{https://cran.r-project.org/web/packages/robis/index.html}{robis} & BIO & R client for the OBIS API \\
\href{https://docs.ropensci.org/EML/}{ropensci/EML} & BIO & Provides support for the serializing and parsing of all low-level EML concepts \\
\href{https://cran.r-project.org/web/packages/taxize/index.html}{taxize} & BIO & Interacts with a suite of web `APIs' for taxonomic tasks, such as getting database specific taxonomic identifiers, verifying species names, getting taxonomic hierarchies, fetching downstream and upstream taxonomic names, getting taxonomic synonyms, converting scientific to common names and vice versa, and more. \\
\href{https://cran.r-project.org/web/packages/worrms/index.html}{worrms} & BIO & Client for \href{http://www.marinespecies.org/}{World Register of Marine Species}. Includes functions for each of the API methods, including searching for names by name, date and common names, searching using external identifiers, fetching synonyms, as well as fetching taxonomic children and taxonomic classification. \\
\href{https://www.rdocumentation.org/packages/Hmisc/versions/4.6-0}{Hmisc} & GEN & Contains many functions useful for data analysis, high-level graphics, utility operations, functions for computing sample size and power, simulation, importing and annotating datasets, imputing missing values, advanced table making, variable clustering, character string manipulation, conversion of R objects to LaTeX and html code, and recoding variables. Particularly check out the \href{https://www.rdocumentation.org/packages/Hmisc/versions/4.6-0/topics/describe}{describe()} function. \\
\href{https://cran.r-project.org/web/packages/lubridate/index.html}{lubridate} & GEN & Functions to work with date-times and time-spans: fast and user friendly parsing of date-time data, extraction and updating of components of a date-time (years, months, days, hours, minutes, and seconds), algebraic manipulation on date-time and time-span objects. \\
\href{https://cran.r-project.org/web/packages/stringr/index.html}{stringr} & GEN & Simple, Consistent Wrappers for Common String Operations \\
\href{https://cran.r-project.org/web/packages/tidyverse/index.html}{tidyverse} & GEN & The `tidyverse' is a set of packages that work in harmony because they share common data representations and `API' design. This package is designed to make it easy to install and load multiple `tidyverse' packages in a single step. \\
\href{https://cran.r-project.org/web/packages/uuid/index.html}{uuid} & GEN & Tools for generating and handling of UUIDs (Universally Unique Identifiers). \\
\bottomrule
\end{longtable}

\hypertarget{python}{%
\section{Python}\label{python}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.28}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.39}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.33}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Type
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://pypi.org/project/metapype/}{metapype} & BIO & A lightweight Python 3 library for generating EML metadata \\
\href{https://python-dwca-reader.readthedocs.io/en/latest/index.html}{python-dwca-reader} & BIO & A simple Python package to read and parse Darwin Core Archive (DwC-A) files, as produced by the GBIF website, the IPT and many other biodiversity informatics tools. \\
\href{https://github.com/iobis/pyworms}{pyworms} & BIO & Python client for the World Register of Marine Species (WoRMS) REST service. \\
\href{https://numpy.org/}{numpy} & GEN & NumPy (Numerical Python) is an open source Python library that's used in almost every field of science and engineering. It's the universal standard for working with numerical data in Python, and it's at the core of the scientific Python and PyData ecosystems. \\
\href{https://pandas.pydata.org/}{pandas} & GEN & pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language. Super helpful when manipulating tabular data! \\
\href{https://docs.python.org/3/library/uuid.html}{uuid} & GEN & This module provides immutable UUID objects (class UUID) and the functions uuid1(), uuid3(), uuid4(), uuid5() for generating version 1, 3, 4, and 5 UUIDs as specified in RFC 4122. Built in -- part of the Python standard library. \\
\href{https://github.com/iobis/obis-qc}{obis-qc} & BIO & Quality checks on occurrence records. Checks \texttt{occurrenceStatus}, \texttt{individualCount}, \texttt{eventDate}, \texttt{decimalLatitude}, \texttt{decimalLongitude}, \texttt{coordinateUncertaintyInMeters}, \texttt{minimumDepthInMeters}, \texttt{maximumDepthInMeters}, \texttt{scientificName}, \texttt{scientificNameID}. Checks from \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4309024/pdf/bau125.pdf}{Vandepitte et al.} flags not implemented: 3, 9, 14, 15, 16, 10, 17, 21-30. \\
\bottomrule
\end{longtable}

\hypertarget{google-sheets}{%
\section{Google Sheets}\label{google-sheets}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.40}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.60}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://dwcaassistant.com/}{Google Sheet DarwinCore Archive Assistant add-on} & Google Sheet add-on which assists the creation of Darwin Core Archives (DwCA) and publising to Zenodo. DwCA's are stored into user's Google Drive and can be downloaded for upload into IPT installations or other software which is able to read DwC-archives. \\
\bottomrule
\end{longtable}

\hypertarget{validators}{%
\section{Validators}\label{validators}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.29}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.71}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Name
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://tools.gbif.org/dwca-validator/}{Darwin Core Archive Validator} & This validator verifies the structural integrity of a Darwin Core Archive. It does not check the data values, such as coordinates, dates or scientific names. \\
\href{https://www.gbif.org/tools/data-validator}{GBIF DATA VALIDATOR} & The GBIF data validator is a service that allows anyone with a GBIF-relevant dataset to receive a report on the syntactical correctness and the validity of the content contained within the dataset. \\
\href{https://www.lifewatch.be/data-services/}{LifeWatch Belgium} & Through this interactive section of the LifeWatch.be portal users can upload their own data using a standard data format, and choose from several web services, models and applications to process the data. \\
\bottomrule
\end{longtable}

\hypertarget{references}{%
\chapter*{References}\label{references}}
\addcontentsline{toc}{chapter}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-R-rmarkdown}{}}%
Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and Richard Iannone. 2021. \emph{Rmarkdown: Dynamic Documents for r}. \url{https://CRAN.R-project.org/package=rmarkdown}.

\leavevmode\vadjust pre{\hypertarget{ref-R-magrittr}{}}%
Bache, Stefan Milton, and Hadley Wickham. 2022. \emph{Magrittr: A Forward-Pipe Operator for r}. \url{https://CRAN.R-project.org/package=magrittr}.

\leavevmode\vadjust pre{\hypertarget{ref-Barbier2017}{}}%
Barbier, Edward B. 2017. {``Marine Ecosystem Services.''} \emph{Current Biology}. Cell Press. \url{https://doi.org/10.1016/j.cub.2017.03.020}.

\leavevmode\vadjust pre{\hypertarget{ref-Benson2018}{}}%
Benson, Abigail, Cassandra M. Brooks, Gabrielle Canonico, Emmett Duffy, Frank Muller-Karger, Heidi M. Sosik, Patricia Miloslavich, and Eduardo Klein. 2018. {``Integrated Observations and Informatics Improve Understanding of Changing Marine Ecosystems.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2018.00428}.

\leavevmode\vadjust pre{\hypertarget{ref-benson_lascala-gruenewald_mcguinn_satterthwaite_beaulieu_biddle_dewitt_mckinzie_montes_moustahfid_etal_2021}{}}%
Benson, Abigail, Diana LaScala-Gruenewald, Robert McGuinn, Erin Satterthwaite, Stace Beaulieu, Mathew Biddle, Lynn deWitt, et al. 2021. {``Biological Observation Data Standardization - a Primer for Data Managers.''} ESIP. \url{https://doi.org/10.6084/m9.figshare.16806712.v1}.

\leavevmode\vadjust pre{\hypertarget{ref-article}{}}%
Benson, Abigail, Tylar Murray, Gabrielle Canonico, Enrique Montes, Frank Muller-Karger, Maria T. Kavanaugh, Joaquin Trinanes, and Lynn M. deWitt. 2021. {``Data Management and Interactive Visualizations for the Evolving Marine Biodiversity Observation Network.''} \emph{Oceanography}. \url{https://doi.org/10.5670/oceanog.2021.220}.

\leavevmode\vadjust pre{\hypertarget{ref-Borer2009}{}}%
Borer, Elizabeth T., Eric W. Seabloom, Matthew B. Jones, and Mark Schildhauer. 2009. {``Some Simple Guidelines for Effective Data Management.''} \emph{Bulletin of the Ecological Society of America} 90 (April). \url{https://doi.org/10.1890/0012-9623-90.2.205}.

\leavevmode\vadjust pre{\hypertarget{ref-Canonico2019}{}}%
Canonico, Gabrielle, Pier Luigi Buttigieg, Enrique Montes, Frank E. Muller-Karger, Carol Stepien, Dawn Wright, Abigail Benson, et al. 2019. {``Global Observational Needs and Resources for Marine Biodiversity.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2019.00367}.

\leavevmode\vadjust pre{\hypertarget{ref-R-worrms}{}}%
Chamberlain, Scott. 2020. \emph{Worrms: World Register of Marine Species (WoRMS) Client}. \url{https://CRAN.R-project.org/package=worrms}.

\leavevmode\vadjust pre{\hypertarget{ref-Crystal-Ornela2021}{}}%
Crystal-Ornelas, Robert, Charuleka Varadharajan, Ben Bond-Lamberty, Kristin Boye, Madison Burrus, Shreyas Cholia, Michael Crow, et al. 2021. {``A Guide to Using GitHub for Developing and Versioning Data Standards and Reporting Formats.''} \emph{Earth and Space Science}, July, e2021EA001797. \url{https://doi.org/10.1029/2021EA001797}.

\leavevmode\vadjust pre{\hypertarget{ref-Davies2021}{}}%
Davies, Neil, John Deck, Eric C Kansa, Sarah Whitcher Kansa, John Kunze, Christopher Meyer, Thomas Orrell, et al. 2021. {``Internet of Samples (iSamples): Toward an Interdisciplinary Cyberinfrastructure for Material Samples.''} \emph{GigaScience} 10: 1--5. \url{https://doi.org/10.1093/gigascience/giab028}.

\leavevmode\vadjust pre{\hypertarget{ref-Djurhuus2020}{}}%
Djurhuus, Anni, Collin J. Closek, Ryan P. Kelly, Kathleen J. Pitz, Reiko P. Michisaki, Hilary A. Starks, Kristine R. Walz, et al. 2020. {``Environmental DNA Reveals Seasonal Shifts and Potential Interactions in a Marine Community.''} \emph{Nature Communications} 11 (December): 1--9. \url{https://doi.org/10.1038/s41467-019-14105-1}.

\leavevmode\vadjust pre{\hypertarget{ref-Duffy2013}{}}%
Duffy, J. Emmett, Linda A. Amaral-Zettler, Daphne G. Fautin, Gustav Paulay, Tatiana A. Rynearson, Heidi M. Sosik, and John J. Stachowicz. 2013. {``Envisioning a Marine Biodiversity Observation Network.''} \emph{BioScience} 63 (May): 350--61. \url{https://doi.org/10.1525/bio.2013.63.5.8}.

\leavevmode\vadjust pre{\hypertarget{ref-Fornwall2012}{}}%
Fornwall, M, R Gisiner, S E Simmons, H Moustahfid, G Canonico, P Halpin, P Goldstein, et al. 2012. {``Expanding Biological Data Standards Development Processes for US IOOS: Visual Line Transect Observing Community for Mammal, Bird, and Turtle Data.''} IOOS. \url{https://www.researchgate.net/publication/255681522}.

\leavevmode\vadjust pre{\hypertarget{ref-lubridate2011}{}}%
Grolemund, Garrett, and Hadley Wickham. 2011. {``Dates and Times Made Easy with {lubridate}.''} \emph{Journal of Statistical Software} 40 (3): 1--25. \url{https://www.jstatsoft.org/v40/i03/}.

\leavevmode\vadjust pre{\hypertarget{ref-Hardisty2019}{}}%
Hardisty, Alex R., William K. Michener, Donat Agosti, Enrique Alonso García, Lucy Bastin, Lee Belbin, Anne Bowser, et al. 2019. {``The Bari Manifesto: An Interoperability Framework for Essential Biodiversity Variables.''} \emph{Ecological Informatics} 49 (January): 22--31. \url{https://doi.org/10.1016/j.ecoinf.2018.11.003}.

\leavevmode\vadjust pre{\hypertarget{ref-Heberling2021}{}}%
Heberling, J Mason, Joseph T Miller, Daniel Noesgaard, Scott B Weingart C, Dmitry Schigel, and Douglas E Soltis. 2021. {``Data Integration Enables Global Biodiversity Synthesis.''} \emph{Proceedings of the National Academy of Sciences of the United States of America}. \url{https://doi.org/10.1073/pnas.2018093118/-/DCSupplemental}.

\leavevmode\vadjust pre{\hypertarget{ref-R-purrr}{}}%
Henry, Lionel, and Hadley Wickham. 2020. \emph{Purrr: Functional Programming Tools}. \url{https://CRAN.R-project.org/package=purrr}.

\leavevmode\vadjust pre{\hypertarget{ref-Hare2016PLOS}{}}%
Jonathan A. Hare, Mark W. Nelson, Wendy E. Morrison. n.d. {``A Vulnerability Assessment of Fish and Invertebrates to Climate Change on the Northeast u.s. Continental Shelf.''} \emph{PLoS ONE} 11 (2): e0146756. \url{https://doi.org/10.1371/journal.pone.0146756}.

\leavevmode\vadjust pre{\hypertarget{ref-Jones2006}{}}%
Jones, Matthew B., Mark P. Schildhauer, O. J. Reichman, and Shawn Bowers. 2006. {``The New Bioinformatics: Integrating Ecological Data from the Gene to the Biosphere.''} \emph{Annual Review of Ecology, Evolution, and Systematics}. \url{https://doi.org/10.1146/annurev.ecolsys.37.091305.110031}.

\leavevmode\vadjust pre{\hypertarget{ref-Kavanaugh2016}{}}%
Kavanaugh, Maria T., Matthew J. Oliver, Francisco P. Chavez, Ricardo M. Letelier, Frank E. Muller-Karger, and Scott C. Doney. 2016. {``Seascapes as a New Vernacular for Pelagic Ocean Monitoring, Management and Conservation.''} \emph{ICES Journal of Marine Science} 73 (July): 1839--50. \url{https://doi.org/10.1093/icesjms/fsw086}.

\leavevmode\vadjust pre{\hypertarget{ref-Kot2010}{}}%
Kot, Connie Y., Ei Fujioka, Lucie J. Hazen, Benjamin D. Best, Andrew J. Read, and Patrick N. Halpin. 2010. {``Spatio-Temporal Gap Analysis of OBIS-SEAMAP Project Data: Assessment and Way Forward.''} \emph{PLoS ONE} 5 (September): 12990. \url{https://doi.org/10.1371/journal.pone.0012990}.

\leavevmode\vadjust pre{\hypertarget{ref-OceanAdapt}{}}%
Lab, Malin Pinksy. n.d. {``OceanAdapt.''} \emph{GitHub}. \url{https://oceanadapt.rutgers.edu/}.

\leavevmode\vadjust pre{\hypertarget{ref-Lamprecht2019}{}}%
Lamprecht, Anna-Lena, Leyla Garcia, Mateusz Kuzak, Carlos Martinez, Ricardo Arcila, Eva Martin Del Pico, Victoria Dominguez Del Angel, et al. 2019. {``Towards FAIR Principles for Research Software.''} Edited by Paul Groth. \emph{Data Science}, 1--23. \url{https://doi.org/10.3233/DS-190026}.

\leavevmode\vadjust pre{\hypertarget{ref-Dornelas2014Science}{}}%
Maria Dornelas, Brian McGill, Nicholas J. Gotelli. 2014. {``Assemblage Time Series Reveal Biodiversity Change but Not Systematic Loss.''} \emph{Science} 344 (6181): 296--99. \url{https://doi.org/10.1126/science.1248484}.

\leavevmode\vadjust pre{\hypertarget{ref-McKenna2021}{}}%
McKenna, Megan F., Simone Baumann-Pickering, Annebelle C. M. Kok, William K. Oestreich, Jeffrey D. Adams, Jack Barkowski, Kurt M. Fristrup, et al. 2021. {``Advancing the Interpretation of Shallow Water Marine Soundscapes.''} \emph{Frontiers in Marine Science} 0 (September): 1426. \url{https://doi.org/10.3389/FMARS.2021.719258}.

\leavevmode\vadjust pre{\hypertarget{ref-Miloslavich2018}{}}%
Miloslavich, Patricia, Nicholas J. Bax, Samantha E. Simmons, Eduardo Klein, Ward Appeltans, Octavio Aburto-Oropeza, Melissa Andersen Garcia, et al. 2018. {``Essential Ocean Variables for Global Sustained Observations of Biodiversity and Ecosystem Changes.''} \emph{Global Change Biology} 24 (June): 2416--33. \url{https://doi.org/10.1111/gcb.14108}.

\leavevmode\vadjust pre{\hypertarget{ref-Montes2020}{}}%
Montes, Enrique, Anni Djurhuus, Frank E. Muller-Karger, Daniel Otis, Christopher R. Kelble, and Maria T. Kavanaugh. 2020. {``Dynamic Satellite Seascapes as a Biogeographic Framework for Understanding Phytoplankton Assemblages in the Florida Keys National Marine Sanctuary, United States.''} \emph{Frontiers in Marine Science} 7 (July): 575. \url{https://doi.org/10.3389/fmars.2020.00575}.

\leavevmode\vadjust pre{\hypertarget{ref-Moustahfid2014}{}}%
Moustahfid, Hassan, and Philip Goldstein. 2014. {``IOOS Biological Data Services Enrollment Procedures.''}

\leavevmode\vadjust pre{\hypertarget{ref-Moustahfid2011}{}}%
Moustahfid, Hassan, Jim Potemra, Philip Goldstein, Roy Mendelssohn, and Annette Desrochers. 2011. {``Making United States Integrated Ocean Observing System (u.s. IOOS) Inclusive of Marine Biological Resources.''} \url{https://www.researchgate.net/publication/254013004}.

\leavevmode\vadjust pre{\hypertarget{ref-R-here}{}}%
Müller, Kirill. 2020. \emph{Here: A Simpler Way to Find Your Files}. \url{https://CRAN.R-project.org/package=here}.

\leavevmode\vadjust pre{\hypertarget{ref-R-tibble}{}}%
Müller, Kirill, and Hadley Wickham. 2021. \emph{Tibble: Simple Data Frames}. \url{https://CRAN.R-project.org/package=tibble}.

\leavevmode\vadjust pre{\hypertarget{ref-Muller-Karger2018b}{}}%
Muller-Karger, Frank E., Erin Hestir, Christiana Ade, Kevin Turpie, Dar A. Roberts, David Siegel, Robert J. Miller, et al. 2018. {``Satellite Sensor Requirements for Monitoring Essential Biodiversity Variables of Coastal Ecosystems.''} \emph{Ecological Applications} 28 (April): 749--60. \url{https://doi.org/10.1002/eap.1682}.

\leavevmode\vadjust pre{\hypertarget{ref-Muller-Karger2018a}{}}%
Muller-Karger, Frank E., Patricia Miloslavich, Nicholas J. Bax, Samantha Simmons, Mark J. Costello, Isabel Sousa Pinto, Gabrielle Canonico, et al. 2018. {``Advancing Marine Biological Observations and Data Requirements of the Complementary Essential Ocean Variables (EOVs) and Essential Biodiversity Variables (EBVs) Frameworks.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2018.00211}.

\leavevmode\vadjust pre{\hypertarget{ref-Neeley2021}{}}%
Neeley, Aimee, Stace E. Beaulieu, Chris Proctor, Ivona Cetinić, Joe Futrelle, Inia Soto Ramos, Heidi M. Sosik, et al. 2021. {``Standards and Practices for Reporting Plankton and Other Particle Observations from Images.''} \url{https://doi.org/10.1575/1912/27377}.

\leavevmode\vadjust pre{\hypertarget{ref-Obrien2021}{}}%
O'Brien, Margaret, Colin A. Smith, Eric R. Sokol, Corinna Gries, Nina Lany, Sydne Record, and Max C. N. Castorani. 2021. {``ecocomDP: A Flexible Data Design Pattern for Ecological Community Survey Data.''} \emph{Ecological Informatics} 64 (September): 101374. \url{https://doi.org/10.1016/J.ECOINF.2021.101374}.

\leavevmode\vadjust pre{\hypertarget{ref-Pooter2017}{}}%
Pooter, Daphnis De, Ward Appeltans, Nicolas Bailly, Sky Bristol, Klaas Deneudt, Menashè Eliezer, Ei Fujioka, et al. 2017. {``Toward a New Data Standard for Combined Marine Biological and Environmental Datasets - Expanding OBIS Beyond Species Occurrences.''} \emph{Biodiversity Data Journal} 5 (January): 10989. \url{https://doi.org/10.3897/BDJ.5.e10989}.

\leavevmode\vadjust pre{\hypertarget{ref-iobis_ebsa}{}}%
Provoost, Pieter. n.d. {``Iobis/Ebsa.''} \emph{GitHub}. \url{https://github.com/iobis/ebsa}.

\leavevmode\vadjust pre{\hypertarget{ref-R-base}{}}%
R Core Team. 2021. \emph{R: A Language and Environment for Statistical Computing}. Vienna, Austria: R Foundation for Statistical Computing. \url{https://www.R-project.org/}.

\leavevmode\vadjust pre{\hypertarget{ref-Ruxfccknagel2015}{}}%
Rücknagel, J., P. Vierkant, R. Ulrich, G. Kloska, E. Schnepf, D. Fichtmüller, E. Reuter, et al. 2015. {``Metadata Schema for the Description of Research Data Repositories. Version 3.0.''} \url{https://doi.org/10.2312/re3.008}.

\leavevmode\vadjust pre{\hypertarget{ref-Rule2019}{}}%
Rule, Adam, Amanda Birmingham, Cristal Zuniga, Ilkay Altintas, Shih-Cheng Huang, Rob Knight, Niema Moshiri, et al. 2019. {``Ten Simple Rules for Writing and Sharing Computational Analyses in Jupyter Notebooks.''} \emph{PLOS Computational Biology} 15 (July): e1007007. \url{https://doi.org/10.1371/JOURNAL.PCBI.1007007}.

\leavevmode\vadjust pre{\hypertarget{ref-Santora2017}{}}%
Santora, Jarrod A., Elliott L. Hazen, Isaac D. Schroeder, Steven J. Bograd, Keith M. Sakuma, and John C. Field. 2017. {``Impacts of Ocean Climate Variability on Biodiversity of Pelagic Forage Species in an Upwelling Ecosystem.''} \emph{Marine Ecology Progress Series} 580 (September): 205--20. \url{https://doi.org/10.3354/meps12278}.

\leavevmode\vadjust pre{\hypertarget{ref-Schmid_Daprano_Jacobson_Sullivan_Briseuxf1o-Avena_Luo_Cowen_2021}{}}%
Schmid, Moritz S, Dominic Daprano, Kyler M Jacobson, Christopher Sullivan, Christian Briseño-Avena, Jessica Y Luo, and Robert K Cowen. 2021. \emph{A Convolutional Neural Network Based High-Throughput Image Classification Pipeline - Code and Documentation to Process Plankton Underwater Imagery Using Local HPC Infrastructure and NSF's XSEDE}. Zenodo. \url{https://doi.org/10.5281/ZENODO.4641158}.

\leavevmode\vadjust pre{\hypertarget{ref-R-lubridate}{}}%
Spinu, Vitalie, Garrett Grolemund, and Hadley Wickham. 2021. \emph{Lubridate: Make Dealing with Dates a Little Easier}. \url{https://CRAN.R-project.org/package=lubridate}.

\leavevmode\vadjust pre{\hypertarget{ref-Taylor2012}{}}%
Taylor, Gordon T., Frank E. Muller-Karger, Robert C. Thunell, Mary I. Scranton, Yrene Astor, Ramon Varela, Luis Troccoli Ghinaglia, et al. 2012. {``Ecosystem Responses in the Southern Caribbean Sea to Global Climate Change.''} \emph{Proceedings of the National Academy of Sciences of the United States of America} 109 (November): 19315--20. \url{https://doi.org/10.1073/pnas.1207514109}.

\leavevmode\vadjust pre{\hypertarget{ref-tittensor2010global}{}}%
Tittensor, Derek P, Camilo Mora, Walter Jetz, Heike K Lotze, Daniel Ricard, Edward Vanden Berghe, and Boris Worm. 2010. {``Global Patterns and Predictors of Marine Biodiversity Across Taxa.''} \emph{Nature} 466 (7310): 1098.

\leavevmode\vadjust pre{\hypertarget{ref-Warren2018}{}}%
Warren, R., J. Price, E. Graham, N. Forstenhaeusler, and J. VanDerWal. 2018. {``The Projected Effect on Insects, Vertebrates, and Plants of Limiting Global Warming to 1.5°c Rather Than 2°c.''} \emph{Science} 360 (May): 791--95. \url{https://doi.org/10.1126/science.aar3646}.

\leavevmode\vadjust pre{\hypertarget{ref-ggplot22016}{}}%
Wickham, Hadley. 2016. \emph{Ggplot2: Elegant Graphics for Data Analysis}. Springer-Verlag New York. \url{https://ggplot2.tidyverse.org}.

\leavevmode\vadjust pre{\hypertarget{ref-R-stringr}{}}%
---------. 2019. \emph{Stringr: Simple, Consistent Wrappers for Common String Operations}. \url{https://CRAN.R-project.org/package=stringr}.

\leavevmode\vadjust pre{\hypertarget{ref-R-forcats}{}}%
---------. 2021a. \emph{Forcats: Tools for Working with Categorical Variables (Factors)}. \url{https://CRAN.R-project.org/package=forcats}.

\leavevmode\vadjust pre{\hypertarget{ref-R-tidyverse}{}}%
---------. 2021b. \emph{Tidyverse: Easily Install and Load the Tidyverse}. \url{https://CRAN.R-project.org/package=tidyverse}.

\leavevmode\vadjust pre{\hypertarget{ref-tidyverse2019}{}}%
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D'Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. {``Welcome to the {tidyverse}.''} \emph{Journal of Open Source Software} 4 (43): 1686. \url{https://doi.org/10.21105/joss.01686}.

\leavevmode\vadjust pre{\hypertarget{ref-R-ggplot2}{}}%
Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen, Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, and Dewey Dunnington. 2021. \emph{Ggplot2: Create Elegant Data Visualisations Using the Grammar of Graphics}. \url{https://CRAN.R-project.org/package=ggplot2}.

\leavevmode\vadjust pre{\hypertarget{ref-R-dplyr}{}}%
Wickham, Hadley, Romain François, Lionel Henry, and Kirill Müller. 2022. \emph{Dplyr: A Grammar of Data Manipulation}. \url{https://CRAN.R-project.org/package=dplyr}.

\leavevmode\vadjust pre{\hypertarget{ref-R-tidyr}{}}%
Wickham, Hadley, and Maximilian Girlich. 2022. \emph{Tidyr: Tidy Messy Data}. \url{https://CRAN.R-project.org/package=tidyr}.

\leavevmode\vadjust pre{\hypertarget{ref-R-readr}{}}%
Wickham, Hadley, Jim Hester, and Jennifer Bryan. 2022. \emph{Readr: Read Rectangular Text Data}. \url{https://CRAN.R-project.org/package=readr}.

\leavevmode\vadjust pre{\hypertarget{ref-Wieczorek2012}{}}%
Wieczorek, John, David Bloom, Robert Guralnick, Stan Blum, Markus Döring, Renato Giovanni, Tim Robertson, and David Vieglais. 2012. {``Darwin Core: An Evolving Community-Developed Biodiversity Data Standard.''} \emph{PLoS ONE} 7 (January): 29715. \url{https://doi.org/10.1371/journal.pone.0029715}.

\leavevmode\vadjust pre{\hypertarget{ref-Wilkinson2016}{}}%
Wilkinson, Mark D., Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. {``The FAIR Guiding Principles for Scientific Data Management and Stewardship.''} \emph{Scientific Data} 3 (March): 1--9. \url{https://doi.org/10.1038/sdata.2016.18}.

\leavevmode\vadjust pre{\hypertarget{ref-knitr2014}{}}%
Xie, Yihui. 2014. {``Knitr: A Comprehensive Tool for Reproducible Research in {R}.''} In \emph{Implementing Reproducible Computational Research}, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC. \url{http://www.crcpress.com/product/isbn/9781466561595}.

\leavevmode\vadjust pre{\hypertarget{ref-xie2015}{}}%
---------. 2015b. \emph{Dynamic Documents with {R} and Knitr}. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. \url{http://yihui.name/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-knitr2015}{}}%
---------. 2015a. \emph{Dynamic Documents with {R} and Knitr}. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://yihui.org/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-bookdown2016}{}}%
---------. 2016. \emph{Bookdown: Authoring Books and Technical Documents with {R} Markdown}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/bookdown}.

\leavevmode\vadjust pre{\hypertarget{ref-R-bookdown}{}}%
---------. 2021a. \emph{Bookdown: Authoring Books and Technical Documents with r Markdown}. \url{https://CRAN.R-project.org/package=bookdown}.

\leavevmode\vadjust pre{\hypertarget{ref-R-knitr}{}}%
---------. 2021b. \emph{Knitr: A General-Purpose Package for Dynamic Report Generation in r}. \url{https://yihui.org/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-rmarkdown2018}{}}%
Xie, Yihui, J. J. Allaire, and Garrett Grolemund. 2018. \emph{R Markdown: The Definitive Guide}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/rmarkdown}.

\leavevmode\vadjust pre{\hypertarget{ref-rmarkdown2020}{}}%
Xie, Yihui, Christophe Dervieux, and Emily Riederer. 2020. \emph{R Markdown Cookbook}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/rmarkdown-cookbook}.

\end{CSLReferences}

\end{document}
